<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to Tidy Transcriptomics • bioceurope2020tidytranscriptomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Tidy Transcriptomics">
<meta property="og:description" content="bioceurope2020tidytranscriptomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bioceurope2020tidytranscriptomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/background.html">Background</a>
</li>
<li>
  <a href="../articles/tidytranscriptomics.html">Workshop</a>
</li>
<li>
  <a href="../articles/solutions.html">Solutions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/bioceurope2020_tidytranscriptomics">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/repository/docker/stemangiola/bioceurope2020_tidytranscriptomics">
    <span class="fa fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tidytranscriptomics_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="tidytranscriptomics_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="tidytranscriptomics_files/anchor-sections-1.0/anchor-sections.js"></script><script src="tidytranscriptomics_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="tidytranscriptomics_files/plotly-binding-4.9.2.1/plotly.js"></script><script src="tidytranscriptomics_files/typedarray-0.1/typedarray.min.js"></script><script src="tidytranscriptomics_files/jquery-1.11.3/jquery.min.js"></script><link href="tidytranscriptomics_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="tidytranscriptomics_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><link href="tidytranscriptomics_files/plotly-htmlwidgets-css-1.52.2/plotly-htmlwidgets.css" rel="stylesheet">
<script src="tidytranscriptomics_files/plotly-main-1.52.2/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to Tidy Transcriptomics</h1>
                        <h4 class="author">Maria Doyle, Peter MacCallum Cancer Centre<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 class="author">Stefano Mangiola, Walter and Eliza Hall Institute<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/bioceurope2020_tidytranscriptomics/blob/master/vignettes/tidytranscriptomics.Rmd"><code>vignettes/tidytranscriptomics.Rmd</code></a></small>
      <div class="hidden name"><code>tidytranscriptomics.Rmd</code></div>

    </div>

    
    
<style type="text/css">
# Formatting for polls
.poll {
background-color: #fff4d4;
}
.poll code {
    background-color: #fff4d4;
}
</style>
<div id="background" class="section level1">
<h1 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h1>
<p>Background information and setup instructions for this workshop can be found <a href="https://stemangiola.github.io/bioceurope2020_tidytranscriptomics/articles/background.html">here</a>.</p>
</div>
<div id="part-1-bulk-rna-seq-with-tidybulk" class="section level1">
<h1 class="hasAnchor">
<a href="#part-1-bulk-rna-seq-with-tidybulk" class="anchor"></a>Part 1 Bulk RNA-seq with tidybulk</h1>
<p>This workshop will present how to perform analysis of RNA sequencing data following the tidy data paradigm <span class="citation">(Wickham and others 2014)</span>. The tidy data paradigm provides a standard way to organise data values within a dataset, where each variable is a column, each observation is a row, and data is manipulated using an easy-to-understand vocabulary. Most importantly, the data structure remains consistent across manipulation and analysis functions.</p>
<p>This can be achieved for RNA sequencing data with the <a href="https://stemangiola.github.io/tidybulk/">tidybulk</a>, <a href="https://stemangiola.github.io/tidyHeatmap">tidyHeatmap</a> <span class="citation">(Mangiola and Papenfuss 2020)</span> and <a href="https://www.tidyverse.org/">tidyverse</a> <span class="citation">(Wickham et al. 2019)</span> packages. The tidybulk package provides a tidy data structure and a modular framework for bulk transcriptional analyses. tidyHeatmap provides a tidy implementation of ComplexHeatmap. These packages are part of the tidytranscriptomics suite that introduces a tidy approach to RNA sequencing data representation and analysis</p>
<div id="acknowledgements" class="section level3">
<h3 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h3>
<p>Some of the material in Part 1 was adapted from an R for RNA sequencing workshop first run <a href="http://combine-australia.github.io/2016-05-11-RNAseq/">here</a>. Use of the airway dataset was inspired by the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2 vignette</a>.</p>
<p><img src="../inst/vignettes/tidybulk_logo.png" width="100px"></p>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA sequencing to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.</p>
<p>There are many steps involved in analysing an RNA sequencing dataset. Sequenced reads are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on in R. While mapping and counting are important and necessary tasks, today we will be starting from the count data and showing how differential expression analysis can be performed in a friendly way using the Bioconductor package, tidybulk.</p>
<p>First, let’s load all the packages we will need to analyse the data.</p>
<p><em>Note: you should load the <em>tidybulk</em> library after the tidyverse core packages for best integration.</em></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load libraries</span>

<span class="co"># dataset</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>

<span class="co"># tidyverse core packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://stringr.tidyverse.org">stringr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>

<span class="co"># tidyverse-friendly packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally">GGally</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org">tidyHeatmap</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidybulk</span><span class="op">)</span></code></pre></div>
<p>Plot settings. Set the colours and theme we will use for our plots.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Use colourblind-friendly colours</span>
<span class="va">friendly_cols</span> <span class="op">&lt;-</span> <span class="fu">dittoSeq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html">dittoColors</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Set theme</span>
<span class="va">custom_theme</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
        panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span><span class="op">)</span>,
        panel.grid.major <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>,
        panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,
        text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">30</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
      <span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<div id="setting-up-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#setting-up-the-data" class="anchor"></a>Setting up the data</h3>
<p>Here we will perform our analysis using the data from the <em>airway</em> package. The airway data comes from the paper by <span class="citation">(Himes et al. 2014)</span>; and it includes 8 samples from human airway smooth muscle cells, from 4 cell lines. For each cell line treated (with dexamethasone) and untreated (negative control) a sample has undergone RNA sequencing and gene counts have been generated.</p>
<p>The airway data is stored as a Bioconductor <em>RangedSummarizedExperiment</em> object. We will convert this object into a tidybulk tibble. A tibble is the tidyverse table format.</p>
<p>In this workshop we will be using the tidyverse pipe <code><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></code>. This ‘pipes’ the output from the command on the left into the command on the right/below. Using the pipe is not essential but it reduces the amount of code we need to write when we have multiple steps (as we’ll see later). It also can make the steps clearer and easier to see. For more details on the pipe see <a href="https://r4ds.had.co.nz/pipes.html">here</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load airway RNA sequencing data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>

<span class="co"># convert to tidybulk tibble</span>
<span class="va">counts_airway</span> <span class="op">&lt;-</span>
  <span class="va">airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">counts_airway</span>
<span class="co">#&gt; # A tibble: 512,816 x 12</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;fct&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 512,806 more rows, and 2 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;</span></code></pre></div>
<p>The <code>counts_airway</code> object contains information about genes and samples, the first column has the Ensembl gene identifier, the second column has the sample identifier and the third column has the gene transcription abundance. The abundance is the number of reads aligning to the gene in each experimental sample. The remaining columns include sample-wise information. The dex column tells us whether the samples are treated or untreated and the cell column tells us what cell line they are from. We can shorten the sample names. We can remove the SRR1039 prefix that’s present in all of them, as shorter names can fit better in some of the plots we will create. We can use <code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code> together with <code><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace()</a></code> to remove the SRR1039 string from the sample column. We can get the gene symbols for these Ensembl gene ids using the Bioconductor annotation package for human, <code>org.Hs.eg.db</code> and add them as a column using <code>mutate</code> again. With tidyverse, all above steps can be linked with the <code><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></code>, as shown below. This has the benefits that * no temporary variables need to be created * less typing is required * the steps can be seen more clearly.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># setup data workflow</span>
<span class="va">counts_tt</span> <span class="op">&lt;-</span>
  <span class="va">airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"SRR1039"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">mapIds</a></span><span class="op">(</span><span class="fu">org.Hs.eg.db</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/org.Hs.eg.db/man/org.Hs.egBASE.html">org.Hs.eg.db</a></span>,
    keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span>,
    keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span>,
    column <span class="op">=</span> <span class="st">"SYMBOL"</span>,
    multiVals <span class="op">=</span> <span class="st">"first"</span>
  <span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 'select()' returned 1:many mapping between keys and columns</span>
<span class="co"># take a look</span>
<span class="va">counts_tt</span>
<span class="co">#&gt; # A tibble: 512,816 x 13</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508         0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508         0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 512,806 more rows, and 3 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;</span></code></pre></div>
<p>From this tidybulk tibble, we can perform differential expression analysis with the tidybulk package.</p>
</div>
</div>
<div id="filtering-lowly-transcribed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-lowly-transcribed-genes" class="anchor"></a>Filtering lowly transcribed genes</h2>
<p>Genes with very low counts across all libraries provide little evidence for differential expression and they can interfere with some of the statistical approximations that are used later in the pipeline. They also add to the multiple testing burden when estimating false discovery rates, reducing power to detect differentially expressed genes. These genes should be filtered out prior to further analysis.</p>
<p>We can perform the filtering using tidybulk <code>keep_abundant</code> or <code>identify_abundant</code>. These functions can use the <em>edgeR</em> <code>filterByExpr</code> function described in <span class="citation">(Law et al. 2016)</span> to automatically identify the genes with adequate abundance for differential expression testing. By default, this will keep genes with ~10 counts in a minimum number of samples, the number of the samples in the smallest group. In this dataset the smallest group size is four (as we have four dex-treated samples versus four untreated). Alternatively, we could use <code>identify_abundant</code> to identify which genes are abundant or not (TRUE/FALSE), rather than just keeping the abundant ones.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Filtering counts</span>
<span class="va">counts_filtered</span> <span class="op">&lt;-</span> <span class="va">counts_tt</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">counts_filtered</span>
<span class="co">#&gt; # A tibble: 127,408 x 14</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 127,398 more rows, and 4 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;</span></code></pre></div>
<p>After running <code>keep_abundant</code> we have a column called <code>.abundant</code> containing TRUE (<code>identify_abundant</code> would have TRUE/FALSE).</p>
</div>
<div id="scaling-counts-to-normalise" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-counts-to-normalise" class="anchor"></a>Scaling counts to normalise</h2>
<p>Scaling of counts, normalisation, is performed to eliminate uninteresting differences between samples due to sequencing depth or composition. A more detailed explanation can be found <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">here</a>. In the tidybulk package the function <code>scale_abundance</code> generates scaled counts, with scaling factors calculated on abundant (filtered) transcripts and applied to all transcripts. We can choose from different normalisation methods. Here we will use the default, edgeR’s trimmed mean of M values (TMM), <span class="citation">(Robinson and Oshlack 2010)</span>. TMM normalisation (and most scaling normalisation methods) scale relative to one sample.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Scaling counts</span>
<span class="va">counts_scaled</span> <span class="op">&lt;-</span> <span class="va">counts_filtered</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">counts_scaled</span>
<span class="co">#&gt; # A tibble: 127,408 x 17</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 127,398 more rows, and 7 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, counts_scaled &lt;dbl&gt;</span></code></pre></div>
<p>After we run <code>scale_abundance</code> we should see some columns have been added at the end. The <code>counts_scaled</code> column contains the scaled counts.</p>
<p>We can visualise the difference of abundance densities before and after scaling. As tidybulk output is compatible with tidyverse, we can simply pipe it into standard tidyverse functions such as <code>filter</code>, <code>pivot_longer</code> and <code>ggplot</code>. We can also take advantage of ggplot’s <code>facet_wrap</code> to easily create multiple plots.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scaled</span> <span class="op">%&gt;%</span>

  <span class="co"># Reshaping</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="st">"counts_scaled"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"source"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plotting</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abundance</span> <span class="op">+</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">source</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-10-1.png" width="70%"></p>
<p>In this dataset the distributions of the counts are not very different to each other before scaling but scaling does make the distributions more similar. If we saw a sample with a very different distribution we may need to investigate it.</p>
<p>As tidybulk smoothly integrates with ggplot2 and other tidyverse packages it can save on typing and make plots easier to generate. Compare the code for creating density plots with tidybulk versus standard base R below (standard code adapted from <span class="citation">(Law et al. 2016)</span>).</p>
<p><strong>tidybulk</strong></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># tidybulk</span>
<span class="va">airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="st">"counts_scaled"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"source"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abundance</span> <span class="op">+</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">source</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><strong>base R using edgeR</strong></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example code, no need to run</span>

<span class="co"># Prepare data set</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span>, group <span class="op">=</span> <span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>, , keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">nsamples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>

<span class="co"># Setup graphics</span>
<span class="va">col</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="va">nsamples</span>, <span class="st">"Paired"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Plot raw counts</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.26</span><span class="op">)</span>, las <span class="op">=</span> <span class="fl">2</span>, main <span class="op">=</span> <span class="st">""</span>, xlab <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span>main <span class="op">=</span> <span class="st">"Counts"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">nsamples</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">den</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">den</span><span class="op">$</span><span class="va">x</span>, <span class="va">den</span><span class="op">$</span><span class="va">y</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend <span class="op">=</span> <span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Run</span>, text.col <span class="op">=</span> <span class="va">col</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span>

<span class="co"># Plot scaled counts</span>
<span class="va">dgList_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">lcpm_n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">dgList_norm</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">lcpm_n</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.26</span><span class="op">)</span>, las <span class="op">=</span> <span class="fl">2</span>, main <span class="op">=</span> <span class="st">""</span>, xlab <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"Counts scaled"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">nsamples</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">den</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">lcpm_n</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">den</span><span class="op">$</span><span class="va">x</span>, <span class="va">den</span><span class="op">$</span><span class="va">y</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend <span class="op">=</span> <span class="va">dgList_norm</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Run</span>, text.col <span class="op">=</span> <span class="va">col</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></code></pre></div>
</div>
<div id="exploratory-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#exploratory-analyses" class="anchor"></a>Exploratory analyses</h2>
<div id="dimensionality-reduction" class="section level3">
<h3 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>Dimensionality reduction</h3>
<p>By far, one of the most important plots we make when we analyse RNA sequencing data are principal-component analysis (PCA) or multi-dimensional scaling (MDS) plots. We reduce the dimensions of the data to identify the greatest sources of variation in the data. A principal components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. If your experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in. It is also an incredibly useful tool for quality control and checking for outliers. We can use the <code>reduce_dimensions</code> function to calculate the dimensions.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get principal components</span>
<span class="va">counts_scal_PCA</span> <span class="op">&lt;-</span>
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span>
<span class="co">#&gt; Fraction of variance explained by the selected principal components</span>
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   `Fraction of variance`    PC</span>
<span class="co">#&gt;                    &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1                  0.353     1</span>
<span class="co">#&gt; 2                  0.312     2</span>
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</span></code></pre></div>
<pre class="poll_4 poll"><code>Poll: What fraction of variance is explained by PC3?
      See ?reduce_dimensions for how to get additional dimensions.</code></pre>
<p>This joins the result to the counts object.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Take a look</span>
<span class="va">counts_scal_PCA</span>
<span class="co">#&gt; # A tibble: 127,408 x 19</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 127,398 more rows, and 9 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, counts_scaled &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;</span></code></pre></div>
<p>For plotting, we can select just the sample-wise information with <code>pivot_sample</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># take a look</span>
<span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 8 x 15</span>
<span class="co">#&gt;   sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt; 1 508    GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 2 509    GSM1275863 N613… trt   untrt SRR1…       126 SRX384346  SRS50…</span>
<span class="co">#&gt; 3 512    GSM1275866 N052… untrt untrt SRR1…       126 SRX384349  SRS50…</span>
<span class="co">#&gt; 4 513    GSM1275867 N052… trt   untrt SRR1…        87 SRX384350  SRS50…</span>
<span class="co">#&gt; 5 516    GSM1275870 N080… untrt untrt SRR1…       120 SRX384353  SRS50…</span>
<span class="co">#&gt; 6 517    GSM1275871 N080… trt   untrt SRR1…       126 SRX384354  SRS50…</span>
<span class="co">#&gt; 7 520    GSM1275874 N061… untrt untrt SRR1…       101 SRX384357  SRS50…</span>
<span class="co">#&gt; 8 521    GSM1275875 N061… trt   untrt SRR1…        98 SRX384358  SRS50…</span>
<span class="co">#&gt; # … with 6 more variables: BioSample &lt;fct&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;</span></code></pre></div>
<p>We can now plot the reduced dimensions.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># PCA plot</span>
<span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, colour <span class="op">=</span> <span class="va">dex</span>, shape <span class="op">=</span> <span class="va">cell</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">sample</span><span class="op">)</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-17-1.png" width="70%"></p>
<p>The samples separate by treatment on PC1 which is what we hope to see. PC2 separates the N080611 cell line from the other samples, indicating a greater difference between that cell line and the others.</p>
</div>
<div id="hierarchical-clustering-with-heatmaps" class="section level3">
<h3 class="hasAnchor">
<a href="#hierarchical-clustering-with-heatmaps" class="anchor"></a>Hierarchical clustering with heatmaps</h3>
<p>An alternative to principal component analysis for examining relationships between samples is using hierarchical clustering. Heatmaps are a nice visualisation to examine hierarchical clustering of your samples. tidybulk has a simple function we can use, <code>keep_variable</code>, to extract the most variable genes which we can then plot with tidyHeatmap.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>

  <span class="co"># extract 500 most variable genes</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_variable-methods.html">keep_variable</a></span><span class="op">(</span>.abundance <span class="op">=</span> <span class="va">counts_scaled</span>, top <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># create heatmap</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap-methods.html">heatmap</a></span><span class="op">(</span>
    .column <span class="op">=</span> <span class="va">sample</span>,
    .row <span class="op">=</span> <span class="va">feature</span>,
    .value <span class="op">=</span> <span class="va">counts_scaled</span>,
    transform <span class="op">=</span> <span class="va">log1p</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/add_tile-methods.html">add_tile</a></span><span class="op">(</span><span class="va">dex</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/add_tile-methods.html">add_tile</a></span><span class="op">(</span><span class="va">cell</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-18-1.png" width="70%"></p>
<p>In the heatmap we can see the samples cluster into two groups, treated and untreated, for three of the cell lines, and the cell line (N080611) again is further away from the others.</p>
<p>Tidybulk enables a simplified way of generating a clustered heatmap of variable genes. Compare the code below for tidybulk versus a base R method.</p>
<p><strong>base R using edgeR</strong></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example code, no need to run</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span>, group <span class="op">=</span> <span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>, , keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">dgList</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span>
<span class="va">select_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">var_genes</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span>
<span class="va">highly_variable_lcpm</span> <span class="op">&lt;-</span> <span class="va">logcounts</span><span class="op">[</span><span class="va">select_var</span>, <span class="op">]</span>
<span class="va">colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#fefada"</span><span class="op">)</span>
<span class="va">col.group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"grey"</span><span class="op">)</span><span class="op">[</span><span class="va">group</span><span class="op">]</span>
<span class="fu">gplots</span><span class="fu">::</span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">highly_variable_lcpm</span>, col <span class="op">=</span> <span class="va">colours</span>, trace <span class="op">=</span> <span class="st">"none"</span>, ColSideColors <span class="op">=</span> <span class="va">col.group</span>, scale <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h2>
<p><em>tidybulk</em> integrates several popular methods for differential transcript abundance testing: the edgeR quasi-likelihood <span class="citation">(Chen, Lun, and Smyth 2016)</span> (tidybulk default method), edgeR likelihood ratio <span class="citation">(McCarthy, Chen, and Smyth 2012)</span>, limma-voom <span class="citation">(Law et al. 2014)</span> and DESeq2 <span class="citation">(Love, Huber, and Anders 2014)</span>. A common question researchers have is which method to choose. With tidybulk we can easily run multiple methods and compare.</p>
<p>We give <code>test_differential_abundance</code> our tidybulk counts object and a formula, specifying the column that contains our groups to be compared. If all our samples were from the same cell line, and there were no additional factors contributing variance such as batch differences, we could use the formula <code>~ dex</code>. However, each treated and untreated sample is from a different cell line so we add the cell line as an additional factor <code>~ dex + cell</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_all</span> <span class="op">&lt;-</span>

  <span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>

  <span class="co"># edgeR QLT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerQLT_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># edgeR LRT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"edgeR_likelihood_ratio"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerLR_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># limma-voom</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"limma_voom"</span>,
    prefix <span class="op">=</span> <span class="st">"voom_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># DESeq2</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"deseq2"</span>,
    prefix <span class="op">=</span> <span class="st">"deseq2_"</span>
  <span class="op">)</span>

<span class="co"># take a look</span>

<span class="va">de_all</span>
<span class="co">#&gt; # A tibble: 127,408 x 41</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 127,398 more rows, and 31 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, counts_scaled &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerQLT_logFC &lt;dbl&gt;, edgerQLT_logCPM &lt;dbl&gt;, edgerQLT_F &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerQLT_PValue &lt;dbl&gt;, edgerQLT_FDR &lt;dbl&gt;, edgerLR_logFC &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_logCPM &lt;dbl&gt;, edgerLR_LR &lt;dbl&gt;, edgerLR_PValue &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_FDR &lt;dbl&gt;, voom_logFC &lt;dbl&gt;, voom_AveExpr &lt;dbl&gt;, voom_t &lt;dbl&gt;,</span>
<span class="co">#&gt; #   voom_P.Value &lt;dbl&gt;, voom_adj.P.Val &lt;dbl&gt;, voom_B &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_baseMean &lt;dbl&gt;, deseq2_log2FoldChange &lt;dbl&gt;, deseq2_lfcSE &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_stat &lt;dbl&gt;, deseq2_pvalue &lt;dbl&gt;, deseq2_padj &lt;dbl&gt;</span></code></pre></div>
<p>This outputs the columns from each method such as log-fold change (logFC), false-discovery rate (FDR) and probability value (p-value). logFC is log2(treated/untreated).</p>
<div id="comparison-of-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#comparison-of-methods" class="anchor"></a>Comparison of methods</h3>
<p>We can visually compare the significance for all methods. We will notice that there is some difference between the methods.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_all</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">edgerQLT_PValue</span>, <span class="va">edgerLR_PValue</span>, <span class="va">voom_P.Value</span>, <span class="va">deseq2_pvalue</span>, <span class="va">feature</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>In order to decide which genes are differentially expressed, we usually take a cut-off of 0.05 on the FDR (or adjusted P value), NOT the raw p-value. This is because we are testing many genes (multiple testing), and the chances of finding differentially expressed genes is very high when you do that many tests. Hence we need to control the false discovery rate, which is the adjusted p-value column in the results table. What this means is that if 100 genes are significant at a 5% false discovery rate, we are willing to accept that 5 will be false positives.</p>
<p>We can count how many differentially expressed genes there are, for example with the edgeR QLT method. We’ll filter on FDR 0.05.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_all</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">edgerQLT_FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>num_de <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html">n_distinct</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 1</span>
<span class="co">#&gt;   num_de</span>
<span class="co">#&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   4967</span></code></pre></div>
<pre class="poll_7 poll"><code>Which method detects the most differentially abundant transcripts, p value adjusted for multiple testing &lt;  0.05 (FDR, adj.P.Val, padj)?</code></pre>
<p><em>Note: Some of the methods produce columns with different names for similar outputs. If you wish to make these consistent you can do that with tidyverse <code>rename</code>. For example, to rename the p value adjusted columns you could run below.</em></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_all</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>deseq2_FDR <span class="op">=</span> <span class="va">deseq2_padj</span>, voom_FDR <span class="op">=</span> <span class="va">voom_adj.P.Val</span><span class="op">)</span></code></pre></div>
<p>We can select some of the transcripts that look different between methods in the plot using the <a href="https://github.com/stemangiola/tidygate"><em>tidygate</em></a> package. We can then investigate these transcripts, for example, by visualising their counts.</p>
<p>With tidygate, we can interactively draw gates to select points we want using <code>gate</code>. We specify which columns we want to plot in the scatterplot, and how many gates we want to draw. We can also specify the opacity if we want to make it easier to see overlapping points.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_gate</span> <span class="op">&lt;-</span>
  <span class="va">de_all</span> <span class="op">%&gt;%</span>

  <span class="fu">tidygate</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidygate/man/gate-methods.html">gate</a></span><span class="op">(</span>
    <span class="va">feature</span>,
    <span class="va">edgerQLT_PValue</span>,
    <span class="va">deseq2_pvalue</span>,
    opacity <span class="op">=</span> <span class="fl">0.3</span>,
    how_many_gates <span class="op">=</span> <span class="fl">2</span>
  <span class="op">)</span></code></pre></div>
<p>We then click to draw gates around the points we want, for example as shown in the screenshot below.</p>
<p><img src="../inst/vignettes/comparison_different_DE_methods_gates.png" width="322"></p>
<p>That will add a column called gate, specifying which gate the points (transcripts) are in.</p>
<pre><code>#&gt; Warning: `gate()` is deprecated as of tidygate 0.3.0.
#&gt; Please use `gate_chr()` instead.
#&gt; This warning is displayed once every 8 hours.
#&gt; Call `lifecycle::last_warnings()` to see where this warning was generated.
#&gt; # A tibble: 127,408 x 42
#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment
#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     
#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 
#&gt; # … with 127,398 more rows, and 32 more variables: Sample &lt;fct&gt;,
#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,
#&gt; #   multiplier &lt;dbl&gt;, counts_scaled &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;,
#&gt; #   edgerQLT_logFC &lt;dbl&gt;, edgerQLT_logCPM &lt;dbl&gt;, edgerQLT_F &lt;dbl&gt;,
#&gt; #   edgerQLT_PValue &lt;dbl&gt;, edgerQLT_FDR &lt;dbl&gt;, edgerLR_logFC &lt;dbl&gt;,
#&gt; #   edgerLR_logCPM &lt;dbl&gt;, edgerLR_LR &lt;dbl&gt;, edgerLR_PValue &lt;dbl&gt;,
#&gt; #   edgerLR_FDR &lt;dbl&gt;, voom_logFC &lt;dbl&gt;, voom_AveExpr &lt;dbl&gt;, voom_t &lt;dbl&gt;,
#&gt; #   voom_P.Value &lt;dbl&gt;, voom_adj.P.Val &lt;dbl&gt;, voom_B &lt;dbl&gt;,
#&gt; #   deseq2_baseMean &lt;dbl&gt;, deseq2_log2FoldChange &lt;dbl&gt;, deseq2_lfcSE &lt;dbl&gt;,
#&gt; #   deseq2_stat &lt;dbl&gt;, deseq2_pvalue &lt;dbl&gt;, deseq2_padj &lt;dbl&gt;, gate &lt;chr&gt;</code></pre>
<p>We can check how many transcripts we’ve got in each gate.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_gate</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">gate</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 x 2</span>
<span class="co">#&gt;   gate      n</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1 0     15905</span>
<span class="co">#&gt; 2 1        12</span>
<span class="co">#&gt; 3 2         9</span></code></pre></div>
<p>We can now select the transcripts from our two gates i.e. more significant in edgeR (gate 1) and more significant in DESeq2 (gate 2) and visualise the counts for each sample in the treated and untreated groups.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_gate</span> <span class="op">%&gt;%</span>

  <span class="co"># Filter for transcripts within the gates</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gate</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Rename for clarity</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>gate <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span>
    <span class="va">gate</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"more in edgeR"</span>,
    <span class="va">gate</span> <span class="op">==</span> <span class="fl">2</span> <span class="op">~</span> <span class="st">"more in DESeq2"</span>,
    <span class="cn">TRUE</span> <span class="op">~</span> <span class="va">gate</span>
  <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Order the plots for the transcripts</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="fu">forcats</span><span class="fu">::</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html">fct_reorder</a></span><span class="op">(</span><span class="va">feature</span>, <span class="va">edgerQLT_PValue</span>, <span class="va">min</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">dex</span>, <span class="va">counts_scaled</span>, color <span class="op">=</span> <span class="va">gate</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">feature</span>, scale <span class="op">=</span> <span class="st">"free_y"</span>, ncol <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>We could also check the log fold changes for these genes and see, for example, that DESeq2 produces a more conservative logFC statistic for the gene <code>ENSG00000104725</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_gate</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">==</span> <span class="st">"ENSG00000104725"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">edgerQLT_logFC</span>, <span class="va">deseq2_log2FoldChange</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 2</span>
<span class="co">#&gt;   edgerQLT_logFC deseq2_log2FoldChange</span>
<span class="co">#&gt;            &lt;dbl&gt;                 &lt;dbl&gt;</span>
<span class="co">#&gt; 1         -0.694                -0.332</span></code></pre></div>
</div>
<div id="single-method" class="section level3">
<h3 class="hasAnchor">
<a href="#single-method" class="anchor"></a>Single method</h3>
<p>If we just wanted to run one differential testing method we could do that. The default method is edgeR quasi-likelihood.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_de</span> <span class="op">&lt;-</span> <span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span><span class="op">)</span>
<span class="co">#&gt; =====================================</span>
<span class="co">#&gt; tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance </span>
<span class="co">#&gt; or adjust_abundance have been calculated. Therefore, it is essential to add covariates </span>
<span class="co">#&gt; such as batch effects (if applicable) in the formula.</span>
<span class="co">#&gt; =====================================</span>
<span class="co">#&gt; tidybulk says: The design column names are "(Intercept), dexuntrt, cellN061011, cellN080611, cellN61311"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span></code></pre></div>
<p>Tidybulk enables a simplified way of performing an RNA sequencing differential expression analysis (with the benefit of smoothly integrating with ggplot2 and other tidyverse packages). Compare the code for a tidybulk edgeR analysis versus standard edgeR below.</p>
<p><strong>standard edgeR</strong></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example code, no need to run</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span>, group <span class="op">=</span> <span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>, , keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span>
<span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">group</span> <span class="op">+</span> <span class="va">cell</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span><span class="op">(</span><span class="va">dgList</span>, <span class="va">design</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFit</a></span><span class="op">(</span><span class="va">dgList</span>, <span class="va">design</span><span class="op">)</span>
<span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="plots-after-testing-for-differentially-expressed" class="section level2">
<h2 class="hasAnchor">
<a href="#plots-after-testing-for-differentially-expressed" class="anchor"></a>Plots after testing for differentially expressed</h2>
<p>We’ll extract the symbols for a few top genes (by P value) to use in some of the plots we will make.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">topgenes_symbols</span> <span class="op">&lt;-</span>
  <span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">PValue</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span></code></pre></div>
<div id="volcano-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#volcano-plots" class="anchor"></a>Volcano plots</h3>
<p>Volcano plots are a useful genome-wide plot for checking that the analysis looks good. Volcano plots enable us to visualise the significance of change (p-value) versus the fold change (logFC). Highly significant genes are towards the top of the plot. We can also colour significant genes (e.g. genes with false-discovery rate &lt; 0.05)</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># volcano plot, minimal</span>
<span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span>, colour <span class="op">=</span> <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10_reverse"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-34-1.png" width="70%"></p>
<p>A more informative plot, integrating some of the packages in tidyverse.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Subset data</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>significant <span class="op">=</span> <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">symbol</span> <span class="op">%in%</span> <span class="va">topgenes_symbols</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>, <span class="st">""</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span>, label <span class="op">=</span> <span class="va">symbol</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">significant</span>, size <span class="op">=</span> <span class="va">significant</span>, alpha <span class="op">=</span> <span class="va">significant</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>

  <span class="co"># Custom scales</span>
  <span class="va">custom_theme</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10_reverse"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#e11f28"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Scale for 'colour' is already present. Adding another scale for 'colour',</span>
<span class="co">#&gt; which will replace the existing scale.</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-35-1.png" width="70%"></p>
</div>
<div id="stripcharts" class="section level3">
<h3 class="hasAnchor">
<a href="#stripcharts" class="anchor"></a>Stripcharts</h3>
<p>Before following up on the differentially expressed genes with further lab work, it is also recommended to have a look at the expression levels of the individual samples for the genes of interest. We can use stripcharts to do this. These will help show if expression is consistent amongst replicates in the groups.</p>
<p>With stripcharts we can see if replicates tend to group together and how the expression compares to the other groups. We’ll also add a box plot to show the distribution. Tidyverse faceting makes it easy to create a plot for each gene.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">strip_chart</span> <span class="op">&lt;-</span>
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span>

  <span class="co"># extract counts for top differentially expressed genes</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">symbol</span> <span class="op">%in%</span> <span class="va">topgenes_symbols</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># make faceted stripchart</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dex</span>, y <span class="op">=</span> <span class="va">counts_scaled</span> <span class="op">+</span> <span class="fl">1</span>, fill <span class="op">=</span> <span class="va">dex</span>, label <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">symbol</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span>

<span class="va">strip_chart</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-36-1.png" width="70%"></p>
</div>
</div>
<div id="interactive-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#interactive-plots" class="anchor"></a>Interactive Plots</h2>
<p>A really nice feature of using tidyverse and ggplot2 is that we can make interactive plots quite easily using the plotly package. This can be very useful for exploring what genes or samples are in the plots. We can make interactive plots directly from our ggplot2 object (strip_chart). Having <code>label</code> in the <code>aes</code> is useful to visualise the identifier of the data point (here the sample id) or other variables when we hover over the plot.</p>
<p>We can also specify which parameters from the <code>aes</code> we want to show up when we hover over the plot with <code>tooltip</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">strip_chart</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/plotly/man/ggplotly.html">ggplotly</a></span><span class="op">(</span>tooltip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"label"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-d519b595363b5fbe86a3" style="width:70%;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-d519b595363b5fbe86a3">{"x":{"data":[{"x":[1,1,1,1],"y":[2.86608249463043,3.24428117804533,3.06839624907108,3.17562691547741],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,1)","legendgroup":"(trt,1)","showlegend":true,"xaxis":"x","yaxis":"y","frame":null},{"x":[1,1,1,1],"y":[3.90486548729367,3.98688675644558,3.81279263895916,4.01655797294455],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,2)","legendgroup":"(trt,2)","showlegend":true,"xaxis":"x2","yaxis":"y","frame":null},{"x":[1,1,1,1],"y":[3.37908997587412,3.66026489164023,3.00017818484172,3.29260776256911],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,3)","legendgroup":"(trt,3)","showlegend":true,"xaxis":"x3","yaxis":"y","frame":null},{"x":[1,1,1,1],"y":[1.82346213237665,2.28135855354415,2.10451752140043,1.96435155901173],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,4)","legendgroup":"(trt,4)","showlegend":true,"xaxis":"x","yaxis":"y2","frame":null},{"x":[1,1,1,1],"y":[3.51387090817402,3.39879151528986,3.29560989216947,3.51160150168158],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,5)","legendgroup":"(trt,5)","showlegend":true,"xaxis":"x2","yaxis":"y2","frame":null},{"x":[1,1,1,1],"y":[3.07311929412043,2.96262137058671,2.77608890608068,3.22821209119327],"hoverinfo":"y","type":"box","fillcolor":"rgba(230,159,0,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(trt,6)","legendgroup":"(trt,6)","showlegend":true,"xaxis":"x3","yaxis":"y2","frame":null},{"x":[2,2,2,2],"y":[2.05771154279207,2.26026869796211,1.92520451632265,2.18242192953093],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,1)","legendgroup":"(untrt,1)","showlegend":true,"xaxis":"x","yaxis":"y","frame":null},{"x":[2,2,2,2],"y":[2.97408516838107,3.17471098929289,3.03966925197695,2.99963058574393],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,2)","legendgroup":"(untrt,2)","showlegend":true,"xaxis":"x2","yaxis":"y","frame":null},{"x":[2,2,2,2],"y":[3.10102428031286,2.79711680070444,2.42359990302169,2.7519404453481],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,3)","legendgroup":"(untrt,3)","showlegend":true,"xaxis":"x3","yaxis":"y","frame":null},{"x":[2,2,2,2],"y":[2.67024519392228,3.04855545571997,2.94533087476182,2.84038005433171],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,4)","legendgroup":"(untrt,4)","showlegend":true,"xaxis":"x","yaxis":"y2","frame":null},{"x":[2,2,2,2],"y":[1.94811580518406,2.08974145700628,2.09104152346256,2.09839137652196],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,5)","legendgroup":"(untrt,5)","showlegend":true,"xaxis":"x2","yaxis":"y2","frame":null},{"x":[2,2,2,2],"y":[0.823513186820998,1.29886003353581,0.346982975464258,0.854869087183615],"hoverinfo":"y","type":"box","fillcolor":"rgba(86,180,233,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"(untrt,6)","legendgroup":"(untrt,6)","showlegend":true,"xaxis":"x3","yaxis":"y2","frame":null},{"x":[1.33184208180755,1.17026127595454,0.879775724187493,1.05522838942707],"y":[3.06820473840114,3.24413166794544,2.86605704960648,3.17545656901837],"text":["counts_scaled + 1:  1170.566926<br />sample: 509","counts_scaled + 1:  1755.016396<br />sample: 513","counts_scaled + 1:   734.653403<br />sample: 517","counts_scaled + 1:  1498.397071<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,1)","legendgroup":"(trt,1)","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.31422493606806,0.683317867480218,0.642182738892734,1.15560900401324],"y":[3.90485943427896,3.98694150358772,3.81284287786758,4.01653520326022],"text":["counts_scaled + 1:  8032.772870<br />sample: 509","counts_scaled + 1:  9702.569369<br />sample: 513","counts_scaled + 1:  6498.193493<br />sample: 517","counts_scaled + 1: 10388.622684<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,2)","legendgroup":"(trt,2)","showlegend":false,"xaxis":"x2","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.38517448753119,0.630680009350181,1.15053665209562,1.00544011890888],"y":[3.66008892340531,3.37898631942836,3.00003245505931,3.2927497243596],"text":["counts_scaled + 1:  4573.670690<br />sample: 509","counts_scaled + 1:  2393.811648<br />sample: 513","counts_scaled + 1:  1000.410370<br />sample: 517","counts_scaled + 1:  1961.587849<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,3)","legendgroup":"(trt,3)","showlegend":false,"xaxis":"x3","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.965158800780773,1.16220476701856,1.07337019145489,0.937946070544422],"y":[1.82353193079864,2.28128869687706,2.10470619141174,1.96427062337089],"text":["counts_scaled + 1:    66.598145<br />sample: 509","counts_scaled + 1:   191.143068<br />sample: 513","counts_scaled + 1:   127.208907<br />sample: 517","counts_scaled + 1:    92.119497<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,4)","legendgroup":"(trt,4)","showlegend":false,"xaxis":"x","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[1.36364174280316,1.38037112150341,0.80313036236912,1.05002427045256],"y":[3.51407543334983,3.39860503339697,3.29551933658418,3.51164909915431],"text":["counts_scaled + 1:  3264.907700<br />sample: 509","counts_scaled + 1:  2504.906475<br />sample: 513","counts_scaled + 1:  1975.194612<br />sample: 517","counts_scaled + 1:  3247.891418<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,5)","legendgroup":"(trt,5)","showlegend":false,"xaxis":"x2","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[1.24832550603896,1.09744865875691,1.3156130220741,0.961803628131747],"y":[3.07300190771145,2.96263663229962,2.77613581482838,3.22819837982535],"text":["counts_scaled + 1:  1183.366564<br />sample: 509","counts_scaled + 1:   917.532319<br />sample: 513","counts_scaled + 1:   597.157520<br />sample: 517","counts_scaled + 1:  1691.266674<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(trt,6)","legendgroup":"(trt,6)","showlegend":false,"xaxis":"x3","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[2.22052855938673,2.27244106829166,2.23252199143171,2.35666903406382],"y":[2.05786838690675,2.2601452048608,1.92538270192939,2.18242013380663],"text":["counts_scaled + 1:   114.211949<br />sample: 508","counts_scaled + 1:   182.082706<br />sample: 512","counts_scaled + 1:    84.179146<br />sample: 516","counts_scaled + 1:   152.202550<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,1)","legendgroup":"(untrt,1)","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.93244104646146,1.77202203534544,1.84849253315479,2.23800888229162],"y":[2.97392662426003,3.17453187692762,2.99973059696454,3.03971728049543],"text":["counts_scaled + 1:   942.074326<br />sample: 508","counts_scaled + 1:  1495.240285<br />sample: 512","counts_scaled + 1:   999.149754<br />sample: 516","counts_scaled + 1:  1095.643463<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,2)","legendgroup":"(untrt,2)","showlegend":false,"xaxis":"x2","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.17595766391605,1.6205116879195,1.89166850764304,1.77951467186213],"y":[3.10106030725397,2.79707746657782,2.42361032290877,2.75205101485134],"text":["counts_scaled + 1:  1261.898082<br />sample: 508","counts_scaled + 1:   626.782411<br />sample: 512","counts_scaled + 1:   265.216111<br />sample: 516","counts_scaled + 1:   564.859511<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,3)","legendgroup":"(untrt,3)","showlegend":false,"xaxis":"x3","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.35830185003579,1.61846830174327,2.19721918497235,1.70584754534066],"y":[2.67040644093716,3.04842772836089,2.94511970779255,2.84023723109791],"text":["counts_scaled + 1:   467.999290<br />sample: 508","counts_scaled + 1:  1118.292612<br />sample: 512","counts_scaled + 1:   881.720371<br />sample: 516","counts_scaled + 1:   692.436662<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,4)","legendgroup":"(untrt,4)","showlegend":false,"xaxis":"x","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[1.75333089511842,2.19105081185698,1.93902964238077,1.69578129723668],"y":[1.94821007372738,2.08966492801999,2.09121894033326,2.09832238236661],"text":["counts_scaled + 1:    88.739260<br />sample: 508","counts_scaled + 1:   122.953659<br />sample: 512","counts_scaled + 1:   123.322274<br />sample: 516","counts_scaled + 1:   125.427099<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,5)","legendgroup":"(untrt,5)","showlegend":false,"xaxis":"x2","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[1.67732372302562,2.19878791477531,1.75186351723969,1.88414064478129],"y":[0.82361201656289,0.855008391113629,0.347031073870837,1.2987733851862],"text":["counts_scaled + 1:     6.660597<br />sample: 508","counts_scaled + 1:     7.159276<br />sample: 512","counts_scaled + 1:     2.223223<br />sample: 516","counts_scaled + 1:    19.900319<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"(untrt,6)","legendgroup":"(untrt,6)","showlegend":false,"xaxis":"x3","yaxis":"y2","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":40.5768157463462,"r":7.30593607305936,"b":44.1700079422781,"l":58.7795765877958},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"xaxis":{"domain":[0,0.32941943900848],"automargin":true,"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["trt","untrt"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["trt","untrt"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-30,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"y2","title":"","hoverformat":".2f"},"annotations":[{"text":"dex","x":0.5,"y":-0.0211346180148144,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"top","annotationType":"axis"},{"text":"counts_scaled + 1","x":-0.0233054616616261,"y":0.5,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"xref":"paper","yref":"paper","textangle":-90,"xanchor":"right","yanchor":"center","annotationType":"axis"},{"text":"CACNB2","x":0.16470971950424,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"DUSP1","x":0.5,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"PDPN","x":0.83529028049576,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"PRSS35","x":0.16470971950424,"y":0.482608570461924,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"SPARCL1","x":0.5,"y":0.482608570461924,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"ZBTB16","x":0.83529028049576,"y":0.482608570461924,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":12.7521793275218},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"dex","x":1.02,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"left","yanchor":"bottom","legendTitle":true}],"yaxis":{"domain":[0.517391429538076,1],"automargin":true,"type":"linear","autorange":false,"range":[0.163504225590244,4.20003672281857],"tickmode":"array","ticktext":["10","100","1000","10000"],"tickvals":[1,2,3,4],"categoryorder":"array","categoryarray":["10","100","1000","10000"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0.517391429538076,"y1":1},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0,"y1":24.4416770444168,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0.517391429538076,"y1":1},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0,"y1":24.4416770444168,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0.517391429538076,"y1":1},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0,"y1":24.4416770444168,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0,"y1":0.482608570461924},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0,"y1":24.4416770444168,"yanchor":0.482608570461924,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0,"y1":0.482608570461924},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0,"y1":24.4416770444168,"yanchor":0.482608570461924,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0,"y1":0.482608570461924},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0,"y1":24.4416770444168,"yanchor":0.482608570461924,"ysizemode":"pixel"}],"xaxis2":{"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["trt","untrt"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["trt","untrt"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-30,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"domain":[0.337247227658187,0.662752772341813],"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"y2","title":"","hoverformat":".2f"},"xaxis3":{"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["trt","untrt"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["trt","untrt"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-30,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"domain":[0.67058056099152,1],"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"y2","title":"","hoverformat":".2f"},"yaxis2":{"type":"linear","autorange":false,"range":[0.163504225590244,4.20003672281857],"tickmode":"array","ticktext":["10","100","1000","10000"],"tickvals":[1,2,3,4],"categoryorder":"array","categoryarray":["10","100","1000","10000"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":12.7521793275218},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":true,"domain":[0,0.482608570461924],"gridcolor":"rgba(235,235,235,1)","gridwidth":0.265670402656704,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":12.7521793275218},"y":0.960675435801739},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"cf135e54385":{"x":{},"y":{},"fill":{},"label":{},"type":"box"},"cf1533c30f3":{"x":{},"y":{},"fill":{},"label":{}}},"cur_data":"cf135e54385","visdat":{"cf135e54385":["function (y) ","x"],"cf1533c30f3":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div id="automatic-bibliography" class="section level2">
<h2 class="hasAnchor">
<a href="#automatic-bibliography" class="anchor"></a>Automatic bibliography</h2>
<p>Tidybulk provides a handy function called <code>get_bibliography</code> that keeps track of the references for the methods used in your tidybulk workflow. The references are in BibTeX format and can be imported into your reference manager.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/get_bibliography-methods.html">get_bibliography</a></span><span class="op">(</span><span class="va">counts_de</span><span class="op">)</span>
<span class="co">#&gt; @article{robinson2010edger,</span>
<span class="co">#&gt;   title={edgeR: a Bioconductor package for differential expression analysis of digital gene expression data},</span>
<span class="co">#&gt;   author={Robinson, Mark D and McCarthy, Davis J and Smyth, Gordon K},</span>
<span class="co">#&gt;   journal={Bioinformatics},</span>
<span class="co">#&gt;   volume={26},</span>
<span class="co">#&gt;   number={1},</span>
<span class="co">#&gt;   pages={139--140},</span>
<span class="co">#&gt;   year={2010},</span>
<span class="co">#&gt;   publisher={Oxford University Press}</span>
<span class="co">#&gt;  }</span>
<span class="co">#&gt; @article{robinson2010scaling,</span>
<span class="co">#&gt;   title={A scaling normalization method for differential expression analysis of RNA-seq data},</span>
<span class="co">#&gt;   author={Robinson, Mark D and Oshlack, Alicia},</span>
<span class="co">#&gt;   journal={Genome biology},</span>
<span class="co">#&gt;   volume={11},</span>
<span class="co">#&gt;   number={3},</span>
<span class="co">#&gt;   pages={1--9},</span>
<span class="co">#&gt;   year={2010},</span>
<span class="co">#&gt;   publisher={BioMed Central}</span>
<span class="co">#&gt;  }</span>
<span class="co">#&gt;   @Manual{,</span>
<span class="co">#&gt;     title = {R: A Language and Environment for Statistical Computing},</span>
<span class="co">#&gt;     author = {{R Core Team}},</span>
<span class="co">#&gt;     organization = {R Foundation for Statistical Computing},</span>
<span class="co">#&gt;     address = {Vienna, Austria},</span>
<span class="co">#&gt;     year = {2020},</span>
<span class="co">#&gt;     url = {https://www.R-project.org/},</span>
<span class="co">#&gt;   }</span>
<span class="co">#&gt; @incollection{smyth2005limma,</span>
<span class="co">#&gt;   title={Limma: linear models for microarray data},</span>
<span class="co">#&gt;   author={Smyth, Gordon K},</span>
<span class="co">#&gt;   booktitle={Bioinformatics and computational biology solutions using R and Bioconductor},</span>
<span class="co">#&gt;   pages={397--420},</span>
<span class="co">#&gt;   year={2005},</span>
<span class="co">#&gt;   publisher={Springer}</span>
<span class="co">#&gt;  }</span></code></pre></div>
</div>
<div id="tidybulk-add-versus-get-modes" class="section level2">
<h2 class="hasAnchor">
<a href="#tidybulk-add-versus-get-modes" class="anchor"></a>Tidybulk ADD versus GET modes</h2>
<p>In the next parts we will see <code>action="get"</code> being used with some tidybulk functions, so we will explain here what it is doing.</p>
<p>Every tidybulk function takes a tidybulk tibble as input, and<br>
* <code>action="add"</code> outputs the new information joined to the original input data frame (default)<br>
* <code>action="get"</code> outputs the new information with only the sample or transcript information, depending on what the analysis is</p>
<p>For example, with <code>action="add"</code> (default), we can add the PCA dimensions to the original data set. So we still have a row for every transcript in every sample.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>
  method <span class="op">=</span> <span class="st">"PCA"</span>,
  action <span class="op">=</span> <span class="st">"add"</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span>
<span class="co">#&gt; Fraction of variance explained by the selected principal components</span>
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   `Fraction of variance`    PC</span>
<span class="co">#&gt;                    &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1                  0.353     1</span>
<span class="co">#&gt; 2                  0.312     2</span>
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</span>
<span class="co">#&gt; # A tibble: 127,408 x 19</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508      2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 127,398 more rows, and 9 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, counts_scaled &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;</span></code></pre></div>
<p>Or with <code>action= "get"</code> we can add the PCA dimensions to the original data set selecting just the sample-wise columns. Note that we now have just one row for every sample.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>
  method <span class="op">=</span> <span class="st">"PCA"</span>,
  action <span class="op">=</span> <span class="st">"get"</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span>
<span class="co">#&gt; Fraction of variance explained by the selected principal components</span>
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   `Fraction of variance`    PC</span>
<span class="co">#&gt;                    &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1                  0.353     1</span>
<span class="co">#&gt; 2                  0.312     2</span>
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</span>
<span class="co">#&gt; # A tibble: 8 x 15</span>
<span class="co">#&gt;   sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt; 1 508    GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 2 509    GSM1275863 N613… trt   untrt SRR1…       126 SRX384346  SRS50…</span>
<span class="co">#&gt; 3 512    GSM1275866 N052… untrt untrt SRR1…       126 SRX384349  SRS50…</span>
<span class="co">#&gt; 4 513    GSM1275867 N052… trt   untrt SRR1…        87 SRX384350  SRS50…</span>
<span class="co">#&gt; 5 516    GSM1275870 N080… untrt untrt SRR1…       120 SRX384353  SRS50…</span>
<span class="co">#&gt; 6 517    GSM1275871 N080… trt   untrt SRR1…       126 SRX384354  SRS50…</span>
<span class="co">#&gt; 7 520    GSM1275874 N061… untrt untrt SRR1…       101 SRX384357  SRS50…</span>
<span class="co">#&gt; 8 521    GSM1275875 N061… trt   untrt SRR1…        98 SRX384358  SRS50…</span>
<span class="co">#&gt; # … with 6 more variables: BioSample &lt;fct&gt;, .abundant &lt;lgl&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;</span></code></pre></div>
</div>
<div id="cell-type-composition-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-composition-analysis" class="anchor"></a>Cell type composition analysis</h2>
<p>If we are sequencing tissue samples, we may want to know what cell types are present and if there are differences in expression between them. <code>tidybulk</code> has a <code>deconvolve_cellularity</code> function that can help us do this.</p>
<p>For this example we will use a subset of the breast cancer dataset from <a href="https://www.cancer.gov/tcga">The Cancer Genome Atlas (TCGA)</a>.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_tidy</span> <span class="op">&lt;-</span>
  <span class="fu">bioceurope2020tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/BRCA.html">BRCA</a></span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="va">patient</span>, <span class="va">transcript</span>, <span class="va">count</span><span class="op">)</span>

<span class="va">BRCA_tidy</span>
<span class="co">#&gt; # A tibble: 198,374 x 5</span>
<span class="co">#&gt;    patient      transcript count  time event_occurred</span>
<span class="co">#&gt;    &lt;fct&gt;        &lt;fct&gt;      &lt;int&gt; &lt;dbl&gt;          &lt;int&gt;</span>
<span class="co">#&gt;  1 TCGA-C8-A275 C1orf112    1571     1              0</span>
<span class="co">#&gt;  2 TCGA-C8-A275 FGR         1873     1              0</span>
<span class="co">#&gt;  3 TCGA-C8-A275 FUCA2       3770     1              0</span>
<span class="co">#&gt;  4 TCGA-C8-A275 GCLC        2738     1              0</span>
<span class="co">#&gt;  5 TCGA-C8-A275 LAS1L       3175     1              0</span>
<span class="co">#&gt;  6 TCGA-C8-A275 ENPP4       1340     1              0</span>
<span class="co">#&gt;  7 TCGA-C8-A275 SEMA3F      5479     1              0</span>
<span class="co">#&gt;  8 TCGA-C8-A275 RAD52        472     1              0</span>
<span class="co">#&gt;  9 TCGA-C8-A275 BAD         1761     1              0</span>
<span class="co">#&gt; 10 TCGA-C8-A275 CD99       14697     1              0</span>
<span class="co">#&gt; # … with 198,364 more rows</span></code></pre></div>
<p>With tidybulk, we can easily infer the proportions of cell types within a tissue using one of several published methods (Cibersort <span class="citation">(Newman et al. 2015)</span>, EPIC <span class="citation">(Racle et al. 2017)</span> and llsr <span class="citation">(Abbas et al. 2009)</span>). Here we will use Cibersort which provides a default signature called LM22 to define the cell types. LM22 contains 547 genes that identify 22 human immune cell types.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type</span> <span class="op">&lt;-</span>
  <span class="va">BRCA_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>action <span class="op">=</span> <span class="st">"get"</span><span class="op">)</span>

<span class="va">BRCA_cell_type</span>
<span class="co">#&gt; # A tibble: 22 x 25</span>
<span class="co">#&gt;    patient  time event_occurred `cibersort: B c… `cibersort: B c…</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;dbl&gt;          &lt;int&gt;            &lt;dbl&gt;            &lt;dbl&gt;</span>
<span class="co">#&gt;  1 TCGA-C…     1              0          0.0325            0     </span>
<span class="co">#&gt;  2 TCGA-A…     1              0          0.0170            0     </span>
<span class="co">#&gt;  3 TCGA-A…     5              0          0.00127           0     </span>
<span class="co">#&gt;  4 TCGA-C…     5              0          0.0372            0     </span>
<span class="co">#&gt;  5 TCGA-P…     5              0          0.00453           0     </span>
<span class="co">#&gt;  6 TCGA-B…  6292              0          0.144             0     </span>
<span class="co">#&gt;  7 TCGA-B…  7106              0          0.0632            0     </span>
<span class="co">#&gt;  8 TCGA-B…  7777              0          0                 0.0142</span>
<span class="co">#&gt;  9 TCGA-B…  8008              0          0.203             0     </span>
<span class="co">#&gt; 10 TCGA-B…  8391              0          0.147             0     </span>
<span class="co">#&gt; # … with 12 more rows, and 20 more variables: `cibersort: Plasma cells` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: T cells CD8` &lt;dbl&gt;, `cibersort: T cells CD4 naive` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: T cells CD4 memory resting` &lt;dbl&gt;, `cibersort: T cells CD4</span>
<span class="co">#&gt; #   memory activated` &lt;dbl&gt;, `cibersort: T cells follicular helper` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: T cells regulatory (Tregs)` &lt;dbl&gt;, `cibersort: T cells gamma</span>
<span class="co">#&gt; #   delta` &lt;dbl&gt;, `cibersort: NK cells resting` &lt;dbl&gt;, `cibersort: NK cells</span>
<span class="co">#&gt; #   activated` &lt;dbl&gt;, `cibersort: Monocytes` &lt;dbl&gt;, `cibersort: Macrophages</span>
<span class="co">#&gt; #   M0` &lt;dbl&gt;, `cibersort: Macrophages M1` &lt;dbl&gt;, `cibersort: Macrophages</span>
<span class="co">#&gt; #   M2` &lt;dbl&gt;, `cibersort: Dendritic cells resting` &lt;dbl&gt;, `cibersort:</span>
<span class="co">#&gt; #   Dendritic cells activated` &lt;dbl&gt;, `cibersort: Mast cells resting` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: Mast cells activated` &lt;dbl&gt;, `cibersort: Eosinophils` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `cibersort: Neutrophils` &lt;dbl&gt;</span></code></pre></div>
<p>Cell type proportions are added to the tibble as new columns. The prefix makes it easy to reshape the data frame if needed, for visualisation or further analyses.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type_long</span> <span class="op">&lt;-</span>
  <span class="va">BRCA_cell_type</span> <span class="op">%&gt;%</span>

  <span class="co"># Reshape</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>
    <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span><span class="op">(</span><span class="st">"cibersort"</span><span class="op">)</span>,
    names_prefix <span class="op">=</span> <span class="st">"cibersort: "</span>,
    names_to <span class="op">=</span> <span class="st">"cell_type"</span>,
    values_to <span class="op">=</span> <span class="st">"proportion"</span>
  <span class="op">)</span>

<span class="va">BRCA_cell_type_long</span>
<span class="co">#&gt; # A tibble: 484 x 5</span>
<span class="co">#&gt;    patient       time event_occurred cell_type                    proportion</span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;dbl&gt;          &lt;int&gt; &lt;chr&gt;                             &lt;dbl&gt;</span>
<span class="co">#&gt;  1 TCGA-C8-A275     1              0 B cells naive                    0.0325</span>
<span class="co">#&gt;  2 TCGA-C8-A275     1              0 B cells memory                   0     </span>
<span class="co">#&gt;  3 TCGA-C8-A275     1              0 Plasma cells                     0.0127</span>
<span class="co">#&gt;  4 TCGA-C8-A275     1              0 T cells CD8                      0     </span>
<span class="co">#&gt;  5 TCGA-C8-A275     1              0 T cells CD4 naive                0.0507</span>
<span class="co">#&gt;  6 TCGA-C8-A275     1              0 T cells CD4 memory resting       0.0299</span>
<span class="co">#&gt;  7 TCGA-C8-A275     1              0 T cells CD4 memory activated     0.0332</span>
<span class="co">#&gt;  8 TCGA-C8-A275     1              0 T cells follicular helper        0     </span>
<span class="co">#&gt;  9 TCGA-C8-A275     1              0 T cells regulatory (Tregs)       0.0326</span>
<span class="co">#&gt; 10 TCGA-C8-A275     1              0 T cells gamma delta              0.0247</span>
<span class="co">#&gt; # … with 474 more rows</span></code></pre></div>
<p>We can plot the proportions of immune cell types for each patient.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type_long</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot proportions</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">patient</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-44-1.png" width="700"></p>
<p>We can visualise the similarity of the tissue composition for the patients by performing a dimensionality reduction on cell type and proportion (rather than on transcript and counts as we did previously).</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type_long</span> <span class="op">%&gt;%</span>

  <span class="co"># Filter cell types with proportion zero in all patients</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">proportion</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>
    <span class="va">patient</span>,
    <span class="va">cell_type</span>,
    <span class="va">proportion</span>,
    method <span class="op">=</span> <span class="st">"PCA"</span>,
    action <span class="op">=</span> <span class="st">"get"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, label <span class="op">=</span> <span class="va">patient</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggrepel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span>
<span class="co">#&gt; Getting the 21 most variable genes</span>
<span class="co">#&gt; Fraction of variance explained by the selected principal components</span>
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   `Fraction of variance`    PC</span>
<span class="co">#&gt;                    &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1                  0.183     1</span>
<span class="co">#&gt; 2                  0.152     2</span>
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-45-1.png" width="700"></p>
<pre class="poll_8 poll"><code>Poll: What is the most abundant cell type overall in BRCA samples?</code></pre>
<p>We can also perform differential tissue composition analyses, similar to how we performed differential transcript abundance analyses. We use tidybulk’s <code>test_differential_cellularity</code> and can perform our analyses using a known factor of interest, such as tumour subtype, or using survival data. Here we use survival data available from TCGA <span class="citation">(Liu et al. 2018)</span>.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival">survival</a></span><span class="op">)</span>

<span class="va">BRCA_tidy_survival</span> <span class="op">&lt;-</span>
  <span class="va">BRCA_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_cellularity-methods.html">test_differential_cellularity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html">Surv</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">event_occurred</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span>

<span class="va">BRCA_tidy_survival</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">.cell_type</span>, <span class="va">p.value</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 22 x 6</span>
<span class="co">#&gt;    .cell_type            p.value cell_type_proport… estimate std.error statistic</span>
<span class="co">#&gt;    &lt;chr&gt;                   &lt;dbl&gt; &lt;list&gt;                &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 NK cells resting       0.0863 &lt;tibble [22 × 6]&gt;     0.330     0.192     1.72 </span>
<span class="co">#&gt;  2 T cells CD4 memory a…  0.0884 &lt;tibble [22 × 6]&gt;     0.313     0.184     1.70 </span>
<span class="co">#&gt;  3 Mast cells resting     0.150  &lt;tibble [22 × 6]&gt;    -0.379     0.264    -1.44 </span>
<span class="co">#&gt;  4 T cells CD4 memory r…  0.161  &lt;tibble [22 × 6]&gt;    -0.865     0.617    -1.40 </span>
<span class="co">#&gt;  5 Macrophages M2         0.164  &lt;tibble [22 × 6]&gt;     0.365     0.262     1.39 </span>
<span class="co">#&gt;  6 T cells gamma delta    0.291  &lt;tibble [22 × 6]&gt;    -0.534     0.506    -1.06 </span>
<span class="co">#&gt;  7 T cells regulatory (…  0.330  &lt;tibble [22 × 6]&gt;    -0.496     0.509    -0.975</span>
<span class="co">#&gt;  8 T cells follicular h…  0.372  &lt;tibble [22 × 6]&gt;    -0.205     0.229    -0.893</span>
<span class="co">#&gt;  9 Macrophages M0         0.520  &lt;tibble [22 × 6]&gt;    -0.248     0.385    -0.643</span>
<span class="co">#&gt; 10 T cells CD4 naive      0.579  &lt;tibble [22 × 6]&gt;     0.113     0.204     0.555</span>
<span class="co">#&gt; # … with 12 more rows</span></code></pre></div>
<p>We can visualise the proportions for the cell types most associated with survival.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_tidy_survival</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">cell_type_proportions</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">.proportion</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">event_occurred</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">.cell_type</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"logit"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span>
<span class="co">#&gt; Warning: Transformation introduced infinite values in continuous y-axis</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-48-1.png" width="700"></p>
</div>
<div id="key-points" class="section level2">
<h2 class="hasAnchor">
<a href="#key-points" class="anchor"></a>Key Points</h2>
<ul>
<li>RNA sequencing data can be represented and analysed in a ‘tidy’ way using tidybulk and the tidyverse</li>
<li>With the modularity offered by piping we don’t need to create variables, unless an object is used more than one. This improves robustness of the code.</li>
<li>The principles of tidy transcriptomics are to interface as much as possible with commonly known manipulation and visualisation tools, rather than creating custom functions.</li>
<li>Some of the key steps in an RNA sequencing analysis are (i) filtering lowly abundant transcripts, (ii) adjusting for differences in sequencing depth and composition, (iii) testing for differential expression</li>
<li>Dimensionality reduction (PCA or MDS) plots are very important for exploring the data</li>
<li>Density plots, volcano plots, strip-charts and heatmaps are useful visualisation tools for evaluating the hypothesis testing.</li>
<li>
<code>tidybulk</code> allows streamlined multi-method analyses</li>
<li>
<code>tidybulk</code> allow easy analyses of cell type composition</li>
<li>Testing for differences in tissue composition between samples is analogous to the testing for differences in transcript abundance</li>
</ul>
</div>
<div id="supplementary" class="section level2">
<h2 class="hasAnchor">
<a href="#supplementary" class="anchor"></a>Supplementary</h2>
<p>Some things we don’t have time to cover in Part 1 of this workshop can be found in the <a href="https://stemangiola.github.io/bioceurope2020_tidytranscriptomics/articles/supplementary.html">Supplementary material</a>.</p>
</div>
</div>
<div id="part-2-bulk-rna-seq-with-tidysummarizedexperiment-and-bioconductor" class="section level1">
<h1 class="hasAnchor">
<a href="#part-2-bulk-rna-seq-with-tidysummarizedexperiment-and-bioconductor" class="anchor"></a>Part 2 Bulk RNA-seq with tidySummarizedExperiment and Bioconductor</h1>
</div>
<div id="tidify-summarizedexperiment-into-tidysummarizedexperiment" class="section level1">
<h1 class="hasAnchor">
<a href="#tidify-summarizedexperiment-into-tidysummarizedexperiment" class="anchor"></a>Tidify <code>SummarizedExperiment</code> into <code>tidySummarizedExperiment</code>
</h1>
<p>This is a SummarizedExperiment object but it is evaluated as tibble. So it is fully compatible both with SummarizedExperiment and tidyverse APIs.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidySummarizedExperiment</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'tidySummarizedExperiment'</span>
<span class="co">#&gt; The following objects are masked from 'package:tidybulk':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     bind_cols, bind_rows, distinct, filter, full_join, group_by,</span>
<span class="co">#&gt;     inner_join, left_join, mutate, nest, rename, right_join, rowwise,</span>
<span class="co">#&gt;     summarise, unnest</span>
<span class="co">#&gt; The following object is masked from 'package:tidyHeatmap':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     pasilla</span>
<span class="co">#&gt; The following object is masked from 'package:plotly':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     plot_ly</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     bind_cols, bind_rows, count</span>
<span class="co">#&gt; The following object is masked from 'package:IRanges':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     slice</span>
<span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     rename</span>
<span class="co">#&gt; The following object is masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     count</span>
<span class="co">#&gt; The following object is masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter</span>
<span class="va">pasilla_tidy</span> <span class="op">&lt;-</span> <span class="va">airway</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span>
<span class="co">#&gt; # A tibble abstraction: 512,816 x 21</span>
<span class="co">#&gt;    sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt;  1 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  2 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  3 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  4 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  5 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  6 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  7 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  8 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  9 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 10 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; # … with 40 more rows, and 12 more variables: BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   transcript &lt;chr&gt;, counts &lt;int&gt;, group &lt;int&gt;, group_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;, strand &lt;fct&gt;,</span>
<span class="co">#&gt; #   exon_id &lt;int&gt;, exon_name &lt;chr&gt;</span></code></pre></div>
<p><strong>But it is a SummarizedExperiment object after all</strong></p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assays</span><span class="op">(</span><span class="va">pasilla_tidy</span><span class="op">)</span>
<span class="co">#&gt; List of length 1</span>
<span class="co">#&gt; names(1): counts</span></code></pre></div>
</div>
<div id="tidyverse-commands" class="section level1">
<h1 class="hasAnchor">
<a href="#tidyverse-commands" class="anchor"></a>Tidyverse commands</h1>
<p>We can use tidyverse commands to explore the tidy SummarizedExperiment object.</p>
<p>We can use <code>slice</code> to choose rows by position, for example to choose the first row.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; # A tibble abstraction: 1 x 21</span>
<span class="co">#&gt;   sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt; 1 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; # … with 12 more variables: BioSample &lt;fct&gt;, transcript &lt;chr&gt;, counts &lt;int&gt;,</span>
<span class="co">#&gt; #   group &lt;int&gt;, group_name &lt;chr&gt;, seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;,</span>
<span class="co">#&gt; #   width &lt;int&gt;, strand &lt;fct&gt;, exon_id &lt;int&gt;, exon_name &lt;chr&gt;</span></code></pre></div>
<p>We can use <code>filter</code> to choose rows by criteria.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">filter</a></span><span class="op">(</span><span class="va">dex</span> <span class="op">==</span> <span class="st">"untrt"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble abstraction: 256,408 x 21</span>
<span class="co">#&gt;    sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt;  1 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  2 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  3 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  4 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  5 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  6 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  7 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  8 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  9 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 10 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; # … with 40 more rows, and 12 more variables: BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   transcript &lt;chr&gt;, counts &lt;int&gt;, group &lt;int&gt;, group_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;, strand &lt;fct&gt;,</span>
<span class="co">#&gt; #   exon_id &lt;int&gt;, exon_name &lt;chr&gt;</span></code></pre></div>
<p>We can use <code>select</code> to choose columns.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 512,816 x 1</span>
<span class="co">#&gt;    sample    </span>
<span class="co">#&gt;    &lt;chr&gt;     </span>
<span class="co">#&gt;  1 SRR1039508</span>
<span class="co">#&gt;  2 SRR1039508</span>
<span class="co">#&gt;  3 SRR1039508</span>
<span class="co">#&gt;  4 SRR1039508</span>
<span class="co">#&gt;  5 SRR1039508</span>
<span class="co">#&gt;  6 SRR1039508</span>
<span class="co">#&gt;  7 SRR1039508</span>
<span class="co">#&gt;  8 SRR1039508</span>
<span class="co">#&gt;  9 SRR1039508</span>
<span class="co">#&gt; 10 SRR1039508</span>
<span class="co">#&gt; # … with 512,806 more rows</span></code></pre></div>
<p>We can use <code>count</code> to count how many rows we have for each sample.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/count.html">count</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 8 x 2</span>
<span class="co">#&gt;   sample         n</span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;int&gt;</span>
<span class="co">#&gt; 1 SRR1039508 64102</span>
<span class="co">#&gt; 2 SRR1039509 64102</span>
<span class="co">#&gt; 3 SRR1039512 64102</span>
<span class="co">#&gt; 4 SRR1039513 64102</span>
<span class="co">#&gt; 5 SRR1039516 64102</span>
<span class="co">#&gt; 6 SRR1039517 64102</span>
<span class="co">#&gt; 7 SRR1039520 64102</span>
<span class="co">#&gt; 8 SRR1039521 64102</span></code></pre></div>
<p>We can use <code>distinct</code> to see what distinct sample information we have.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">distinct</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">dex</span>, <span class="va">cell</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 8 x 3</span>
<span class="co">#&gt;   sample     cell    dex  </span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;fct&gt;   &lt;fct&gt;</span>
<span class="co">#&gt; 1 SRR1039508 N61311  untrt</span>
<span class="co">#&gt; 2 SRR1039509 N61311  trt  </span>
<span class="co">#&gt; 3 SRR1039512 N052611 untrt</span>
<span class="co">#&gt; 4 SRR1039513 N052611 trt  </span>
<span class="co">#&gt; 5 SRR1039516 N080611 untrt</span>
<span class="co">#&gt; 6 SRR1039517 N080611 trt  </span>
<span class="co">#&gt; 7 SRR1039520 N061011 untrt</span>
<span class="co">#&gt; 8 SRR1039521 N061011 trt</span></code></pre></div>
<p>We could use <code>rename</code> to rename a column. For example, to modify the type column name.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">rename</a></span><span class="op">(</span>treatment_status<span class="op">=</span><span class="va">dex</span><span class="op">)</span>
<span class="co">#&gt; # A tibble abstraction: 512,816 x 21</span>
<span class="co">#&gt;    sample SampleName cell  treatment_status albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt;            &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 SRR10… GSM1275862 N613… untrt            untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 40 more rows, and 13 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   transcript &lt;chr&gt;, counts &lt;int&gt;, group &lt;int&gt;, group_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;, strand &lt;fct&gt;,</span>
<span class="co">#&gt; #   exon_id &lt;int&gt;, exon_name &lt;chr&gt;</span></code></pre></div>
<p>We could use <code>mutate</code> to create a column. For example, we could create a new type column that contains single and paired instead of single_end and paired_end.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>type<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"SRR1039"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble abstraction: 512,816 x 22</span>
<span class="co">#&gt;    sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt;  1 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  2 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  3 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  4 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  5 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  6 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  7 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  8 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  9 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 10 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; # … with 40 more rows, and 13 more variables: BioSample &lt;fct&gt;, type &lt;chr&gt;,</span>
<span class="co">#&gt; #   transcript &lt;chr&gt;, counts &lt;int&gt;, group &lt;int&gt;, group_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;, strand &lt;fct&gt;,</span>
<span class="co">#&gt; #   exon_id &lt;int&gt;, exon_name &lt;chr&gt;</span></code></pre></div>
<p>We could use <code>unite</code> to combine multiple columns into a single column.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tidyr-methods.html">unite</a></span><span class="op">(</span><span class="st">"combined"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">dex</span>, <span class="va">cell</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble abstraction: 512,816 x 20</span>
<span class="co">#&gt;    sample SampleName combined albut Run   avgLength Experiment Sample BioSample</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;fct&gt;      &lt;chr&gt;    &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt;  &lt;fct&gt;    </span>
<span class="co">#&gt;  1 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt;  2 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt;  3 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt;  4 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt;  5 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt;  6 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt;  7 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt;  8 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt;  9 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt; 10 SRR10… GSM1275862 untrt_N… untrt SRR1…       126 SRX384345  SRS50… SAMN0242…</span>
<span class="co">#&gt; # … with 40 more rows, and 11 more variables: transcript &lt;chr&gt;, counts &lt;int&gt;,</span>
<span class="co">#&gt; #   group &lt;int&gt;, group_name &lt;chr&gt;, seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;,</span>
<span class="co">#&gt; #   width &lt;int&gt;, strand &lt;fct&gt;, exon_id &lt;int&gt;, exon_name &lt;chr&gt;</span></code></pre></div>
<p>We can also combine commands with the tidyverse pipe <code><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></code>.</p>
<p>For example, we could combine <code>group_by</code> and <code>summarise</code> to get the total counts for each sample.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">group_by</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">summarise</a></span><span class="op">(</span>total_counts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; `summarise()` ungrouping output (override with `.groups` argument)</span>
<span class="co">#&gt; # A tibble: 8 x 2</span>
<span class="co">#&gt;   sample     total_counts</span>
<span class="co">#&gt;   &lt;chr&gt;             &lt;int&gt;</span>
<span class="co">#&gt; 1 SRR1039508     20637971</span>
<span class="co">#&gt; 2 SRR1039509     18809481</span>
<span class="co">#&gt; 3 SRR1039512     25348649</span>
<span class="co">#&gt; 4 SRR1039513     15163415</span>
<span class="co">#&gt; 5 SRR1039516     24448408</span>
<span class="co">#&gt; 6 SRR1039517     30818215</span>
<span class="co">#&gt; 7 SRR1039520     19126151</span>
<span class="co">#&gt; 8 SRR1039521     21164133</span></code></pre></div>
<p>We could combine <code>group_by</code>, <code>mutate</code> and <code>filter</code> to get the transcripts with mean count &gt; 0.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">group_by</a></span><span class="op">(</span><span class="va">seqnames</span> <span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>mean_count<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">filter</a></span><span class="op">(</span><span class="va">mean_count</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 512,816 x 22</span>
<span class="co">#&gt; # Groups:   seqnames [1]</span>
<span class="co">#&gt;    sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt;  1 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  2 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  3 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  4 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  5 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  6 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  7 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  8 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt;  9 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 10 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; # … with 512,806 more rows, and 13 more variables: BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   transcript &lt;chr&gt;, counts &lt;int&gt;, group &lt;int&gt;, group_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;, strand &lt;fct&gt;,</span>
<span class="co">#&gt; #   exon_id &lt;int&gt;, exon_name &lt;chr&gt;, mean_count &lt;dbl&gt;</span></code></pre></div>
<div id="plotting" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting" class="anchor"></a>Plotting</h2>
<p>We can treat <code>pasilla_tidy</code> as a normal tibble for plotting. Here we plot the distribution of counts per sample.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group<span class="op">=</span><span class="va">sample</span>, color<span class="op">=</span><span class="va">`dex`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/plot1-1.png" width="700"></p>
</div>
</div>
<div id="part-3-single-cell-rna-seq-with-tidysinglecellexperiment" class="section level1">
<h1 class="hasAnchor">
<a href="#part-3-single-cell-rna-seq-with-tidysinglecellexperiment" class="anchor"></a>Part 3 Single-cell RNA-seq with tidySingleCellExperiment</h1>
<p><a href="https://stemangiola.github.io/bioceurope2020_tidytranscriptomics/articles/background.html#differences-between-bulk-and-single-cell-rna-sequencing-1">Background information</a></p>
<p>In Part 1 we showed how we can study the cell-type composition of a biological sample using bulk RNA sequencing. Single cell sequencing enables a more direct estimation of cell-type composition and gives greater resolution. For bulk RNA sequencing we need to infer the cell types using the abundance of transcripts in the whole sample, with single-cell RNA sequencing we can directly measure the transcripts in each cell and then classify the cells into cell types.</p>
<div id="introduction-to-tidyseurat" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-tidyseurat" class="anchor"></a>Introduction to tidyseurat</h2>
<p>Seurat is a very popular analysis toolkit for single cell RNA sequencing data <span class="citation">(Butler et al. 2018; Stuart et al. 2019)</span>. tidyseurat provides a bridge between the Seurat single-cell package and the tidyverse. It creates an invisible layer that enables viewing the Seurat object as a tidyverse tibble, and provides Seurat-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># load additional libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stemangiola.github.io/bioceurope2020_tidytranscriptomics/">bioceurope2020tidytranscriptomics</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://stringr.tidyverse.org">stringr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org">igraph</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR">SingleR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex">celldex</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eclarke/ggbeeswarm">ggbeeswarm</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidySingleCellExperiment</span><span class="op">)</span></code></pre></div>
<p>The single-cell RNA sequencing data used here is 3000 cells in total, subsetted from 20 samples from 10 peripheral blood mononuclear cell (pbmc) datasets. The datasets are from GSE115189/SRR7244582 <span class="citation">(Freytag et al. 2018)</span>, SRR11038995 [<span class="citation">Cai et al. (2020)</span>, SCP345 (singlecell.broadinstitute.org), SCP424 <span class="citation">(Ding et al. 2020)</span>, SCP591 <span class="citation">(Karagiannis et al. 2020)</span> and 10x-derived 6K and 8K datasets (support.10xgenomics.com/).</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">bioceurope2020tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/pbmc.html">pbmc</a></span>

<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>
<span class="co">#&gt; [1] 51958  3000</span></code></pre></div>
<div id="create-tidyseurat" class="section level3">
<h3 class="hasAnchor">
<a href="#create-tidyseurat" class="anchor"></a>Create <code>tidyseurat</code>
</h3>
<p>This is a seurat object but it is evaluated as tibble. So it is fully compatible both with Seurat and tidyverse APIs.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_tidy</span> <span class="op">&lt;-</span> <span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_tidy</span>
<span class="co">#&gt; # A tibble abstraction: 3,000 x 8</span>
<span class="co">#&gt;    cell    file      orig.ident nCount_RNA nFeature_RNA  S.Score G2M.Score ident</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;           &lt;dbl&gt;        &lt;int&gt;    &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;</span>
<span class="co">#&gt;  1 CCAGTC… ../data/… SeuratPro…       3421          979 -5.42e-2  -0.0107  G1   </span>
<span class="co">#&gt;  2 ATGAGC… ../data/… SeuratPro…       2752          898 -5.01e-2  -0.00416 G1   </span>
<span class="co">#&gt;  3 TATGAA… ../data/… SeuratPro…       2114          937 -2.95e-5  -0.0229  G1   </span>
<span class="co">#&gt;  4 CATATA… ../data/… SeuratPro…       3122         1086 -6.65e-2  -0.0488  G1   </span>
<span class="co">#&gt;  5 GAGGCA… ../data/… SeuratPro…       2341          957 -3.74e-3   0.0241  G2M  </span>
<span class="co">#&gt;  6 AGCTGC… ../data/… SeuratPro…       5472         1758 -5.88e-2   0.00241 G2M  </span>
<span class="co">#&gt;  7 TGATTA… ../data/… SeuratPro…       1258          542 -2.51e-2  -0.0269  G1   </span>
<span class="co">#&gt;  8 ACGAAG… ../data/… SeuratPro…       7683         1926 -1.33e-1  -0.116   G1   </span>
<span class="co">#&gt;  9 CGGCAT… ../data/… SeuratPro…       3500         1092 -6.87e-2  -0.0622  G1   </span>
<span class="co">#&gt; 10 ATAGCG… ../data/… SeuratPro…       3092          974 -1.24e-2  -0.0271  G1   </span>
<span class="co">#&gt; # … with 2,990 more rows</span></code></pre></div>
<p><strong>But it is a Seurat object after all</strong></p>
<p>So it can be interacted with using <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingleCellExperiment commands</a> such as <code>assayNames</code>.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assayNames</a></span><span class="op">(</span><span class="va">pbmc_tidy</span><span class="op">)</span> <span class="co"># from SummarizedExperiment</span>
<span class="co">#&gt; [1] "counts"    "logcounts"</span></code></pre></div>
</div>
</div>
<div id="polish-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#polish-the-data" class="anchor"></a>Polish the data</h2>
<p>We can also interact with our object as we do with any tidyverse tibble. In this case we want to polish an annotation column. We will extract the sample and dataset names from the file name column into separate columns.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_tidy</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>

  <span class="co"># Extract sample and group</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">extract</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"sample"</span>, <span class="st">"../data/([a-zA-Z0-9_]+)/outs.+"</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Extract data source</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">extract</a></span><span class="op">(</span><span class="va">file</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dataset"</span>, <span class="st">"groups"</span><span class="op">)</span>, <span class="st">"../data/([a-zA-Z0-9_]+)_([0-9])/outs.+"</span><span class="op">)</span>

<span class="va">pbmc_tidy</span>
<span class="co">#&gt; # A tibble abstraction: 3,000 x 10</span>
<span class="co">#&gt;    cell  dataset groups sample orig.ident nCount_RNA nFeature_RNA  S.Score</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;           &lt;dbl&gt;        &lt;int&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt;  1 CCAG… GSE115… 1      GSE11… SeuratPro…       3421          979 -5.42e-2</span>
<span class="co">#&gt;  2 ATGA… GSE115… 2      GSE11… SeuratPro…       2752          898 -5.01e-2</span>
<span class="co">#&gt;  3 TATG… GSE115… 2      GSE11… SeuratPro…       2114          937 -2.95e-5</span>
<span class="co">#&gt;  4 CATA… GSE115… 1      GSE11… SeuratPro…       3122         1086 -6.65e-2</span>
<span class="co">#&gt;  5 GAGG… GSE115… 2      GSE11… SeuratPro…       2341          957 -3.74e-3</span>
<span class="co">#&gt;  6 AGCT… GSE115… 1      GSE11… SeuratPro…       5472         1758 -5.88e-2</span>
<span class="co">#&gt;  7 TGAT… GSE115… 1      GSE11… SeuratPro…       1258          542 -2.51e-2</span>
<span class="co">#&gt;  8 ACGA… GSE115… 1      GSE11… SeuratPro…       7683         1926 -1.33e-1</span>
<span class="co">#&gt;  9 CGGC… GSE115… 1      GSE11… SeuratPro…       3500         1092 -6.87e-2</span>
<span class="co">#&gt; 10 ATAG… GSE115… 1      GSE11… SeuratPro…       3092          974 -1.24e-2</span>
<span class="co">#&gt; # … with 2,990 more rows, and 2 more variables: G2M.Score &lt;dbl&gt;, ident &lt;fct&gt;</span></code></pre></div>
</div>
<div id="quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h2>
<p>A key quality control step performed in single cell analysis is assessment of the proportion of mitochondrial transcripts. A high mitochondrial count can indicate the cell is dead or dying and is thus poor quality. A powerful combination of tools we can use with tidyseurat is tidyverse’s <a href="https://rstudio-education.github.io/tidyverse-cookbook/transform-tables.html#nest-a-data-frame">nest</a> and <code>map</code>. This enables us to easily perform independent analyses on subsets of the data. For example, we can nest by dataset and use map to iterate the analysis of mitochondrial content across each dataset.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get chromosomal location for gene symbols</span>
<span class="va">location</span> <span class="op">&lt;-</span> <span class="fu">mapIds</span><span class="op">(</span>
  <span class="va">EnsDb.Hsapiens.v86</span>,
  keys <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">pbmc_tidy</span><span class="op">)</span>,
  column <span class="op">=</span> <span class="st">"SEQNAME"</span>,
  keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>
<span class="op">)</span>
<span class="co">#&gt; Warning: Unable to map 5313 of 51958 requested IDs.</span>

<span class="va">pbmc_tidy</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>

  <span class="co"># Grouping - nesting by dataset</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">dataset</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>mitochondrion_info <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span>
    <span class="va">data</span>,
    <span class="op">~</span> <span class="co"># Calculating mitochondrial statistics</span>
      <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html">perCellQCMetrics</a></span><span class="op">(</span><span class="va">.x</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Mito <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">location</span> <span class="op">==</span> <span class="st">"MT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
      
      <span class="co"># Converting to tibble</span>
      <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span> <span class="op">%&gt;%</span>

      <span class="co"># Label cells with high mitochondrial content</span>
      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>high_mitochondrion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html">isOutlier</a></span><span class="op">(</span><span class="va">subsets_Mito_percent</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Join the information</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html">map2</a></span><span class="op">(</span>
    <span class="va">data</span>, <span class="va">mitochondrion_info</span>,
    <span class="op">~</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">.x</span>, <span class="va">.y</span>, by <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span>
  <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">mitochondrion_info</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<p>We can use tidyverse to reshape the data and create beeswarm plots to visualise the mitochondrial content.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>

  <span class="co"># Reshaping</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">detected</span>, <span class="va">sum</span>, <span class="va">subsets_Mito_percent</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="va">dataset</span>, y <span class="op">=</span> <span class="va">value</span>,
    color <span class="op">=</span> <span class="va">high_mitochondrion</span>,
    alpha <span class="op">=</span> <span class="va">high_mitochondrion</span>,
    size <span class="op">=</span> <span class="va">high_mitochondrion</span>
  <span class="op">)</span><span class="op">)</span> <span class="op">+</span>

  <span class="co"># Plotting</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_quasirandom.html">geom_quasirandom</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">name</span>, scale <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span>

  <span class="co"># Customisation</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#e11f28"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">50</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-69-1.png" width="700"></p>
<p>In the faceted plot, detected is number of genes in each of the 10 datasets, sum is total counts.</p>
<pre class="poll_5 poll"><code>What is the ID of the cell with the highest mitochondrial relative content?</code></pre>
<p>We filter out cells with high mitochondrial content.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_tidy</span> <span class="op">&lt;-</span> <span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">high_mitochondrion</span><span class="op">)</span></code></pre></div>
</div>
<div id="scaling-and-integrating" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-and-integrating" class="anchor"></a>Scaling and Integrating</h2>
<p>We need to integrate the different datasets to adjust for technical variability between them. For integration we have chosen 10x_8K (dataset number 3) as reference because it is relatively high quality.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Scaling within each dataset</span>
<span class="va">pbmc_tidy</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>

  <span class="co"># Define batch</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>batch <span class="op">=</span> <span class="va">dataset</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">batch</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Normalisation   </span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html">multiBatchNorm</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html">fastMNN</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h2>
<p>Besides PCA which is a linear dimensionality reduction, we can apply neighbour aware methods such as UMAP, to better define locally similar cells. We can calculate the first 3 UMAP dimensions using the Seurat framework.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Perform PCA with scater</span>
<span class="va">pbmc_tidy</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span>ncomponents <span class="op">=</span> <span class="fl">3</span>, dimred<span class="op">=</span><span class="st">"corrected"</span><span class="op">)</span> <span class="co"># from scater</span></code></pre></div>
<p>And we can plot the output as a 3D plot using plotly. With this we can check if the datasets are clustering independently, indicating a batch effect, or if they are mixed together and cells are clustering by cell type.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/plot_ly.html">plot_ly</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP1`</span>,
    y <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP2`</span>,
    z <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP3`</span>,
    colors <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,
    color <span class="op">=</span> <span class="op">~</span><span class="va">batch</span>
  <span class="op">)</span></code></pre></div>
<p><img src="../inst/vignettes/plotly_1.png" width="278"></p>
</div>
<div id="identify-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h2>
<p>We proceed with cluster identification with Seurat. FindNeighbors constructs a Shared Nearest Neighbor (SNN) Graph for a given dataset. The k-nearest neighbours of each cell are first determined. FindClusters identifies clusters of cells by a shared nearest neighbour (SNN).</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Assign clusters to the 'colLabels'</span>
<span class="co"># of the SingleCellExperiment object</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html">colLabels</a></span><span class="op">(</span><span class="va">pbmc_tidy</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="co"># from SingleCellExperiment</span>
  <span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span><span class="op">(</span>use.dimred<span class="op">=</span><span class="st">"corrected"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># from scran - shared nearest neighbour</span>
  <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%$%</span> <span class="co"># from igraph</span>
  <span class="va">membership</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Reorder columns</span>
<span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">label</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble abstraction: 2,899 x 11</span>
<span class="co">#&gt;    cell  label batch corrected1 corrected2 corrected3 corrected4 corrected5</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;fct&gt; &lt;int&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 CCAG… 5         1   -0.0643      0.334     -0.0725    -0.0560    0.0739 </span>
<span class="co">#&gt;  2 ATGA… 3         1   -0.178      -0.0166    -0.122     -0.122    -0.0233 </span>
<span class="co">#&gt;  3 TATG… 6         1   -0.0603      0.0357     0.127     -0.293    -0.0128 </span>
<span class="co">#&gt;  4 CATA… 11        1   -0.0760      0.282     -0.121     -0.0546    0.00283</span>
<span class="co">#&gt;  5 GAGG… 2         1   -0.0757      0.0424     0.0977    -0.259    -0.0248 </span>
<span class="co">#&gt;  6 AGCT… 4         1   -0.117       0.0242    -0.0729    -0.159    -0.0170 </span>
<span class="co">#&gt;  7 TGAT… 1         1   -0.146       0.0752    -0.143     -0.0488   -0.0915 </span>
<span class="co">#&gt;  8 ACGA… 1         1   -0.119       0.108     -0.127     -0.0505   -0.0624 </span>
<span class="co">#&gt;  9 CGGC… 5         1   -0.00941     0.407     -0.0466    -0.0803    0.0523 </span>
<span class="co">#&gt; 10 ATAG… 5         1   -0.0561      0.367     -0.105     -0.0589    0.0609 </span>
<span class="co">#&gt; # … with 2,889 more rows, and 3 more variables: UMAP1 &lt;dbl&gt;, UMAP2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   UMAP3 &lt;dbl&gt;</span></code></pre></div>
<p>Now we can interrogate the object as if it was a regular tibble data frame.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">label</span><span class="op">)</span>
<span class="co">#&gt; tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 11 x 2</span>
<span class="co">#&gt;    label     n</span>
<span class="co">#&gt;    &lt;fct&gt; &lt;int&gt;</span>
<span class="co">#&gt;  1 1       367</span>
<span class="co">#&gt;  2 2       460</span>
<span class="co">#&gt;  3 3       727</span>
<span class="co">#&gt;  4 4       554</span>
<span class="co">#&gt;  5 5       339</span>
<span class="co">#&gt;  6 6       276</span>
<span class="co">#&gt;  7 7        21</span>
<span class="co">#&gt;  8 8        74</span>
<span class="co">#&gt;  9 9        41</span>
<span class="co">#&gt; 10 10       14</span>
<span class="co">#&gt; 11 11       26</span></code></pre></div>
<p>And we can plot the clusters as a 3D plot using plotly. This time we are colouring by estimated cluster labels to visually check the cluster labels.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/plot_ly.html">plot_ly</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP1`</span>,
    y <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP2`</span>,
    z <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP3`</span>,
    colors <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,
    color <span class="op">=</span> <span class="op">~</span><span class="va">label</span>
  <span class="op">)</span></code></pre></div>
<p><img src="../inst/vignettes/plotly_2.png" width="261"></p>
</div>
<div id="cell-type-classification" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-classification" class="anchor"></a>Cell type classification</h2>
<div id="manual-cell-type-classification" class="section level3">
<h3 class="hasAnchor">
<a href="#manual-cell-type-classification" class="anchor"></a>Manual cell type classification</h3>
<p>We can identify cluster markers (genes) using Seurat. As example, we are selecting the top 10 for each cluster. The function DoHeatmap builds an heatmap of those gene markers across cells.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Identify top 10 markers per cluster</span>
<span class="va">marker_genes</span> <span class="op">&lt;-</span>
    <span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers</a></span><span class="op">(</span>groups<span class="op">=</span><span class="va">pbmc_tidy</span><span class="op">$</span><span class="va">label</span>, assay.type <span class="op">=</span> <span class="st">"reconstructed"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">.x</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Plot heatmap</span>
<span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span>                                  <span class="co"># from scater</span>
    features<span class="op">=</span><span class="va">marker_genes</span>, 
    columns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">pbmc_tidy</span><span class="op">$</span><span class="va">label</span><span class="op">)</span>, 
    colour_columns_by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"label"</span><span class="op">)</span>, 
    exprs_values  <span class="op">=</span> <span class="st">"reconstructed"</span>
  <span class="op">)</span> </code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-74-1.png" width="700"></p>
</div>
<div id="automatic-cell-type-classification" class="section level3">
<h3 class="hasAnchor">
<a href="#automatic-cell-type-classification" class="anchor"></a>Automatic cell type classification</h3>
<p>We can infer cell type identities (e.g. T cell) using <em>SingleR</em> <span class="citation">(Aran et al. 2019)</span> and manipulate the output using tidyverse. SingleR <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SingleR/inst/doc/SingleR.html">accepts any log-normalised transcript abundance matrix</a></p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Reference cell types</span>
<span class="va">blueprint</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/datasets.html">BlueprintEncodeData</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; using temporary cache /tmp/RtmpOrjw2F/BiocFileCache</span>
<span class="co">#&gt; snapshotDate(): 2020-10-02</span>
<span class="co">#&gt; see ?celldex and browseVignettes('celldex') for documentation</span>
<span class="co">#&gt; downloading 1 resources</span>
<span class="co">#&gt; retrieving 1 resource</span>
<span class="co">#&gt; loading from cache</span>
<span class="co">#&gt; see ?celldex and browseVignettes('celldex') for documentation</span>
<span class="co">#&gt; downloading 1 resources</span>
<span class="co">#&gt; retrieving 1 resource</span>
<span class="co">#&gt; loading from cache</span>

<span class="va">cell_type</span> <span class="op">&lt;-</span>
    
  <span class="co"># extracting counts from SingleCellExperiment object</span>
  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">pbmc_tidy</span><span class="op">)</span><span class="op">$</span><span class="va">reconstructed</span><span class="op">%&gt;%</span>

    <span class="co"># SingleR</span>
  <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html">SingleR</a></span><span class="op">(</span>
    ref <span class="op">=</span> <span class="va">blueprint</span>,
    labels <span class="op">=</span> <span class="va">blueprint</span><span class="op">$</span><span class="va">label.main</span>,
    clusters <span class="op">=</span> <span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">label</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
    
    <span class="co"># Formatting results</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">first.labels</span><span class="op">)</span></code></pre></div>
<p>We join the cluster label dataframe to our tidyseurat tibble.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Join cell type info</span>
<span class="va">pbmc_tidy</span> <span class="op">&lt;-</span>
  <span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">cell_type</span>, by <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span>

<span class="co"># Reorder columns</span>
<span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble abstraction: 2,899 x 12</span>
<span class="co">#&gt;    cell  first.labels batch label corrected1 corrected2 corrected3 corrected4</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;        &lt;int&gt; &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 CCAG… Monocytes        1 5       -0.0643      0.334     -0.0725    -0.0560</span>
<span class="co">#&gt;  2 ATGA… Skeletal mu…     1 3       -0.178      -0.0166    -0.122     -0.122 </span>
<span class="co">#&gt;  3 TATG… NK cells         1 6       -0.0603      0.0357     0.127     -0.293 </span>
<span class="co">#&gt;  4 CATA… Monocytes        1 11      -0.0760      0.282     -0.121     -0.0546</span>
<span class="co">#&gt;  5 GAGG… NK cells         1 2       -0.0757      0.0424     0.0977    -0.259 </span>
<span class="co">#&gt;  6 AGCT… Fibroblasts      1 4       -0.117       0.0242    -0.0729    -0.159 </span>
<span class="co">#&gt;  7 TGAT… B-cells          1 1       -0.146       0.0752    -0.143     -0.0488</span>
<span class="co">#&gt;  8 ACGA… B-cells          1 1       -0.119       0.108     -0.127     -0.0505</span>
<span class="co">#&gt;  9 CGGC… Monocytes        1 5       -0.00941     0.407     -0.0466    -0.0803</span>
<span class="co">#&gt; 10 ATAG… Monocytes        1 5       -0.0561      0.367     -0.105     -0.0589</span>
<span class="co">#&gt; # … with 2,889 more rows, and 4 more variables: corrected5 &lt;dbl&gt;, UMAP1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   UMAP2 &lt;dbl&gt;, UMAP3 &lt;dbl&gt;</span></code></pre></div>
<p>We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_tidy</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">first.labels</span><span class="op">)</span>
<span class="co">#&gt; tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 11 x 3</span>
<span class="co">#&gt;    label first.labels        n</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;           &lt;int&gt;</span>
<span class="co">#&gt;  1 1     B-cells           367</span>
<span class="co">#&gt;  2 10    B-cells            14</span>
<span class="co">#&gt;  3 11    Monocytes          26</span>
<span class="co">#&gt;  4 2     NK cells          460</span>
<span class="co">#&gt;  5 3     Skeletal muscle   727</span>
<span class="co">#&gt;  6 4     Fibroblasts       554</span>
<span class="co">#&gt;  7 5     Monocytes         339</span>
<span class="co">#&gt;  8 6     NK cells          276</span>
<span class="co">#&gt;  9 7     Smooth muscle      21</span>
<span class="co">#&gt; 10 8     Neutrophils        74</span>
<span class="co">#&gt; 11 9     Macrophages        41</span></code></pre></div>
<pre class="poll_6 poll"><code>How many cells are classified differently by SingleR when analyses are done by cluster (as above) or by cell (omitting the argument `clusters`). Tip: you can answer this question without creating any variable, using left_join.</code></pre>
</div>
</div>
<div id="pseudobulk-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#pseudobulk-analyses" class="anchor"></a>Pseudobulk analyses</h2>
<p>It is sometime useful to aggregate cell-wise transcript abundance into pseudobulk samples. It is possible to explore data and perform hypothesis testing with tools and data-source that we are more familiar with. For example, we can use edgeR in tidybulk to perform differential expression testing. For more details on pseudobulk analysis see <a href="https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html">here</a>.</p>
<div id="data-exploration-using-pseudobulk-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#data-exploration-using-pseudobulk-samples" class="anchor"></a>Data exploration using pseudobulk samples</h3>
<p>To do this, we will load a helper function called <code>aggregate_cells</code> from <a href="https://gist.github.com/stemangiola/ebce5a3f931b298611b56680d39c073c">GitHub Gist</a> and use it to create a group for each sample.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load aggregate function</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/source_gist.html">source_gist</a></span><span class="op">(</span><span class="st">"889e4089a0e5ca2649d3c9164e3b542e"</span><span class="op">)</span>

<span class="co"># Aggregate</span>
<span class="va">pbmc_bulk</span> <span class="op">=</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidy.html">tidy</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">aggregate_cells</span><span class="op">(</span><span class="va">file</span><span class="op">)</span> 

<span class="va">pbmc_bulk</span>
<span class="co">#&gt; # A tibble: 1,039,160 x 5</span>
<span class="co">#&gt;    file                transcript  abundance_counts abundance_logcou… orig.ident</span>
<span class="co">#&gt;    &lt;chr&gt;               &lt;chr&gt;                  &lt;dbl&gt;             &lt;dbl&gt; &lt;chr&gt;     </span>
<span class="co">#&gt;  1 ../data/GSE115189_… DDX11L1                    0                 0 SeuratPro…</span>
<span class="co">#&gt;  2 ../data/GSE115189_… WASH7P                     0                 0 SeuratPro…</span>
<span class="co">#&gt;  3 ../data/GSE115189_… MIR6859-1                  0                 0 SeuratPro…</span>
<span class="co">#&gt;  4 ../data/GSE115189_… MIR1302-2HG                0                 0 SeuratPro…</span>
<span class="co">#&gt;  5 ../data/GSE115189_… MIR1302-2                  0                 0 SeuratPro…</span>
<span class="co">#&gt;  6 ../data/GSE115189_… FAM138A                    0                 0 SeuratPro…</span>
<span class="co">#&gt;  7 ../data/GSE115189_… OR4G4P                     0                 0 SeuratPro…</span>
<span class="co">#&gt;  8 ../data/GSE115189_… OR4G11P                    0                 0 SeuratPro…</span>
<span class="co">#&gt;  9 ../data/GSE115189_… OR4F5                      0                 0 SeuratPro…</span>
<span class="co">#&gt; 10 ../data/GSE115189_… CICP27                     0                 0 SeuratPro…</span>
<span class="co">#&gt; # … with 1,039,150 more rows</span></code></pre></div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pbmc_bulk</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Tidybulk operations</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="va">file</span>, <span class="va">transcript</span>, <span class="va">abundance_counts</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">tidybulk</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">tidybulk</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; No group or design set. Assuming all samples belong to one group.</span>
<span class="co">#&gt; # A tibble: 1,039,160 x 9</span>
<span class="co">#&gt;    file  transcript abundance_counts abundance_logco… orig.ident .abundant   TMM</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;                 &lt;dbl&gt;            &lt;dbl&gt; &lt;chr&gt;      &lt;lgl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 ../d… DDX11L1                   0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt;  2 ../d… WASH7P                    0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt;  3 ../d… MIR6859-1                 0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt;  4 ../d… MIR1302-2…                0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt;  5 ../d… MIR1302-2                 0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt;  6 ../d… FAM138A                   0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt;  7 ../d… OR4G4P                    0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt;  8 ../d… OR4G11P                   0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt;  9 ../d… OR4F5                     0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt; 10 ../d… CICP27                    0                0 SeuratPro… FALSE     0.785</span>
<span class="co">#&gt; # … with 1,039,150 more rows, and 2 more variables: multiplier &lt;dbl&gt;,</span>
<span class="co">#&gt; #   abundance_counts_scaled &lt;dbl&gt;</span></code></pre></div>
</div>
</div>
<div id="key-points-1" class="section level2">
<h2 class="hasAnchor">
<a href="#key-points-1" class="anchor"></a>Key Points</h2>
<ul>
<li>Some basic steps of a single-cell RNA sequencing analysis are dimensionality reduction, cluster identification and cell type classification</li>
<li>
<code>tidyseurat</code> is an invisible layer that operates on a <code>Seurat</code> object and enables us to visualise and manipulate data as if it were a tidy data frame.</li>
<li>
<code>tidyseurat</code> object is a <code>Seurat object</code> so it can be used with any <code>Seurat</code> compatible method</li>
<li>
<code>tidySingleCellExperiment</code> is an invisible layer that operates on a <code>SingleCellExperiment</code> object and enables us to visualise and manipulate data as if it were a tidy data frame.</li>
<li>
<code>tidySingleCellExperiment</code> object is a <code>SingleCellExperiment object</code> so it can be used with any <code>SingleCellExperiment</code> compatible method</li>
</ul>
</div>
</div>
<div id="contributing" class="section level1">
<h1 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h1>
<p>If you want to suggest improvements for this workshop or ask questions, you can do so as described <a href="https://github.com/stemangiola/bioceurope2020_tidytranscriptomics/blob/master/CONTRIBUTING.md">here</a>.</p>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>Record package and version information with <code>sessionInfo</code></p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] tidySingleCellExperiment_1.0.0         </span>
<span class="co">#&gt;  [2] ggbeeswarm_0.6.0                       </span>
<span class="co">#&gt;  [3] celldex_1.0.0                          </span>
<span class="co">#&gt;  [4] EnsDb.Hsapiens.v86_2.99.0              </span>
<span class="co">#&gt;  [5] ensembldb_2.14.0                       </span>
<span class="co">#&gt;  [6] AnnotationFilter_1.14.0                </span>
<span class="co">#&gt;  [7] GenomicFeatures_1.42.1                 </span>
<span class="co">#&gt;  [8] AnnotationDbi_1.52.0                   </span>
<span class="co">#&gt;  [9] scuttle_1.0.3                          </span>
<span class="co">#&gt; [10] SingleR_1.4.0                          </span>
<span class="co">#&gt; [11] batchelor_1.6.2                        </span>
<span class="co">#&gt; [12] igraph_1.2.6                           </span>
<span class="co">#&gt; [13] scran_1.18.2                           </span>
<span class="co">#&gt; [14] scater_1.18.3                          </span>
<span class="co">#&gt; [15] SingleCellExperiment_1.12.0            </span>
<span class="co">#&gt; [16] purrr_0.3.4                            </span>
<span class="co">#&gt; [17] bioceurope2020tidytranscriptomics_0.3.0</span>
<span class="co">#&gt; [18] tidySummarizedExperiment_1.0.0         </span>
<span class="co">#&gt; [19] survival_3.2-7                         </span>
<span class="co">#&gt; [20] tidybulk_1.2.0                         </span>
<span class="co">#&gt; [21] tidyHeatmap_1.1.5                      </span>
<span class="co">#&gt; [22] GGally_2.0.0                           </span>
<span class="co">#&gt; [23] ggrepel_0.8.2                          </span>
<span class="co">#&gt; [24] plotly_4.9.2.1                         </span>
<span class="co">#&gt; [25] ggplot2_3.3.2                          </span>
<span class="co">#&gt; [26] stringr_1.4.0                          </span>
<span class="co">#&gt; [27] readr_1.4.0                            </span>
<span class="co">#&gt; [28] tidyr_1.1.2                            </span>
<span class="co">#&gt; [29] dplyr_1.0.2                            </span>
<span class="co">#&gt; [30] tibble_3.0.4                           </span>
<span class="co">#&gt; [31] airway_1.10.0                          </span>
<span class="co">#&gt; [32] SummarizedExperiment_1.20.0            </span>
<span class="co">#&gt; [33] Biobase_2.50.0                         </span>
<span class="co">#&gt; [34] GenomicRanges_1.42.0                   </span>
<span class="co">#&gt; [35] GenomeInfoDb_1.26.2                    </span>
<span class="co">#&gt; [36] IRanges_2.24.0                         </span>
<span class="co">#&gt; [37] S4Vectors_0.28.1                       </span>
<span class="co">#&gt; [38] BiocGenerics_0.36.0                    </span>
<span class="co">#&gt; [39] MatrixGenerics_1.2.0                   </span>
<span class="co">#&gt; [40] matrixStats_0.57.0                     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] rappdirs_0.3.1                rtracklayer_1.50.0           </span>
<span class="co">#&gt;   [3] ragg_0.4.0                    bit64_4.0.5                  </span>
<span class="co">#&gt;   [5] knitr_1.30                    irlba_2.3.3                  </span>
<span class="co">#&gt;   [7] DelayedArray_0.16.0           data.table_1.13.4            </span>
<span class="co">#&gt;   [9] RCurl_1.98-1.2                generics_0.1.0               </span>
<span class="co">#&gt;  [11] preprocessCore_1.52.0         callr_3.5.1                  </span>
<span class="co">#&gt;  [13] cowplot_1.1.0                 usethis_2.0.0                </span>
<span class="co">#&gt;  [15] RSQLite_2.2.1                 bit_4.0.4                    </span>
<span class="co">#&gt;  [17] xml2_1.3.2                    httpuv_1.5.4                 </span>
<span class="co">#&gt;  [19] assertthat_0.2.1              viridis_0.5.1                </span>
<span class="co">#&gt;  [21] xfun_0.19                     hms_0.5.3                    </span>
<span class="co">#&gt;  [23] evaluate_0.14                 promises_1.1.1               </span>
<span class="co">#&gt;  [25] fansi_0.4.1                   progress_1.2.2               </span>
<span class="co">#&gt;  [27] dbplyr_2.0.0                  DBI_1.1.0                    </span>
<span class="co">#&gt;  [29] geneplotter_1.68.0            htmlwidgets_1.5.3            </span>
<span class="co">#&gt;  [31] reshape_0.8.8                 ellipsis_0.3.1               </span>
<span class="co">#&gt;  [33] RSpectra_0.16-0               crosstalk_1.1.0.1            </span>
<span class="co">#&gt;  [35] backports_1.2.1               annotate_1.68.0              </span>
<span class="co">#&gt;  [37] biomaRt_2.46.0                sparseMatrixStats_1.2.0      </span>
<span class="co">#&gt;  [39] vctrs_0.3.5                   remotes_2.2.0                </span>
<span class="co">#&gt;  [41] Cairo_1.5-12.2                withr_2.3.0                  </span>
<span class="co">#&gt;  [43] GenomicAlignments_1.26.0      prettyunits_1.1.1            </span>
<span class="co">#&gt;  [45] dittoSeq_1.2.2                cluster_2.1.0                </span>
<span class="co">#&gt;  [47] ExperimentHub_1.16.0          lazyeval_0.2.2               </span>
<span class="co">#&gt;  [49] crayon_1.3.4                  genefilter_1.72.0            </span>
<span class="co">#&gt;  [51] edgeR_3.32.0                  pkgconfig_2.0.3              </span>
<span class="co">#&gt;  [53] labeling_0.4.2                vipor_0.4.5                  </span>
<span class="co">#&gt;  [55] pkgload_1.1.0                 ProtGenerics_1.22.0          </span>
<span class="co">#&gt;  [57] devtools_2.3.2                rlang_0.4.9                  </span>
<span class="co">#&gt;  [59] lifecycle_0.2.0               BiocFileCache_1.14.0         </span>
<span class="co">#&gt;  [61] rsvd_1.0.3                    AnnotationHub_2.22.0         </span>
<span class="co">#&gt;  [63] rprojroot_2.0.2               Matrix_1.2-18                </span>
<span class="co">#&gt;  [65] boot_1.3-25                   beeswarm_0.2.3               </span>
<span class="co">#&gt;  [67] ggridges_0.5.2                GlobalOptions_0.1.2          </span>
<span class="co">#&gt;  [69] processx_3.4.5                pheatmap_1.0.12              </span>
<span class="co">#&gt;  [71] png_0.1-7                     viridisLite_0.3.0            </span>
<span class="co">#&gt;  [73] rjson_0.2.20                  bitops_1.0-6                 </span>
<span class="co">#&gt;  [75] Biostrings_2.58.0             blob_1.2.1                   </span>
<span class="co">#&gt;  [77] DelayedMatrixStats_1.12.1     shape_1.4.5                  </span>
<span class="co">#&gt;  [79] beachmat_2.6.2                scales_1.1.1                 </span>
<span class="co">#&gt;  [81] memoise_1.1.0                 magrittr_2.0.1               </span>
<span class="co">#&gt;  [83] plyr_1.8.6                    zlibbioc_1.36.0              </span>
<span class="co">#&gt;  [85] compiler_4.0.3                dqrng_0.2.1                  </span>
<span class="co">#&gt;  [87] RColorBrewer_1.1-2            tidygate_0.3.2               </span>
<span class="co">#&gt;  [89] clue_0.3-58                   DESeq2_1.30.0                </span>
<span class="co">#&gt;  [91] Rsamtools_2.6.0               cli_2.2.0                    </span>
<span class="co">#&gt;  [93] XVector_0.30.0                ps_1.5.0                     </span>
<span class="co">#&gt;  [95] tidyselect_1.1.0              stringi_1.5.3                </span>
<span class="co">#&gt;  [97] forcats_0.5.0                 textshaping_0.2.1            </span>
<span class="co">#&gt;  [99] yaml_2.2.1                    BiocSingular_1.6.0           </span>
<span class="co">#&gt; [101] askpass_1.1                   locfit_1.5-9.4               </span>
<span class="co">#&gt; [103] grid_4.0.3                    tools_4.0.3                  </span>
<span class="co">#&gt; [105] circlize_0.4.11               bluster_1.0.0                </span>
<span class="co">#&gt; [107] gridExtra_2.3                 farver_2.0.3                 </span>
<span class="co">#&gt; [109] digest_0.6.27                 BiocManager_1.30.10          </span>
<span class="co">#&gt; [111] FNN_1.1.3                     shiny_1.5.0                  </span>
<span class="co">#&gt; [113] Rcpp_1.0.5                    broom_0.7.2                  </span>
<span class="co">#&gt; [115] BiocVersion_3.12.0            later_1.1.0.1                </span>
<span class="co">#&gt; [117] org.Hs.eg.db_3.12.0           httr_1.4.2                   </span>
<span class="co">#&gt; [119] ComplexHeatmap_2.6.2          colorspace_2.0-0             </span>
<span class="co">#&gt; [121] XML_3.99-0.5                  fs_1.5.0                     </span>
<span class="co">#&gt; [123] splines_4.0.3                 uwot_0.1.9                   </span>
<span class="co">#&gt; [125] statmod_1.4.35                pkgdown_1.6.1                </span>
<span class="co">#&gt; [127] sessioninfo_1.1.1             systemfonts_0.3.2            </span>
<span class="co">#&gt; [129] xtable_1.8-4                  jsonlite_1.7.2               </span>
<span class="co">#&gt; [131] testthat_3.0.0                R6_2.5.0                     </span>
<span class="co">#&gt; [133] pillar_1.4.7                  htmltools_0.5.0              </span>
<span class="co">#&gt; [135] mime_0.9                      glue_1.4.2                   </span>
<span class="co">#&gt; [137] fastmap_1.0.1                 BiocParallel_1.24.1          </span>
<span class="co">#&gt; [139] BiocNeighbors_1.8.2           class_7.3-17                 </span>
<span class="co">#&gt; [141] interactiveDisplayBase_1.28.0 pkgbuild_1.1.0               </span>
<span class="co">#&gt; [143] utf8_1.1.4                    lattice_0.20-41              </span>
<span class="co">#&gt; [145] ResidualMatrix_1.0.0          curl_4.3                     </span>
<span class="co">#&gt; [147] openssl_1.4.3                 limma_3.46.0                 </span>
<span class="co">#&gt; [149] rmarkdown_2.5                 desc_1.2.0                   </span>
<span class="co">#&gt; [151] munsell_0.5.0                 e1071_1.7-4                  </span>
<span class="co">#&gt; [153] GetoptLong_1.0.4              GenomeInfoDbData_1.2.4       </span>
<span class="co">#&gt; [155] gtable_0.3.0</span></code></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-abbas2009deconvolution">
<p>Abbas, Alexander R, Kristen Wolslegel, Dhaya Seshasayee, Zora Modrusan, and Hilary F Clark. 2009. “Deconvolution of Blood Microarray Data Identifies Cellular Activation Patterns in Systemic Lupus Erythematosus.” <em>PloS One</em> 4 (7): e6098.</p>
</div>
<div id="ref-aran2019reference">
<p>Aran, Dvir, Agnieszka P Looney, Leqian Liu, Esther Wu, Valerie Fong, Austin Hsu, Suzanna Chak, et al. 2019. “Reference-Based Analysis of Lung Single-Cell Sequencing Reveals a Transitional Profibrotic Macrophage.” <em>Nature Immunology</em> 20 (2): 163–72.</p>
</div>
<div id="ref-butler2018integrating">
<p>Butler, Andrew, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. 2018. “Integrating Single-Cell Transcriptomic Data Across Different Conditions, Technologies, and Species.” <em>Nature Biotechnology</em> 36 (5): 411–20.</p>
</div>
<div id="ref-Cai2020">
<p>Cai, Yi, Youchao Dai, Yejun Wang, Qianqing Yang, Jiubiao Guo, Cailing Wei, Weixin Chen, et al. 2020. “Single-Cell Transcriptomics of Blood Reveals a Natural Killer Cell Subset Depletion in Tuberculosis.” <em>EBioMedicine</em> 53 (March): 102686. <a href="https://doi.org/10.1016/j.ebiom.2020.102686">https://doi.org/10.1016/j.ebiom.2020.102686</a>.</p>
</div>
<div id="ref-chen2016reads">
<p>Chen, Yunshun, Aaron TL Lun, and Gordon K Smyth. 2016. “From Reads to Genes to Pathways: Differential Expression Analysis of Rna-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.” <em>F1000Research</em> 5.</p>
</div>
<div id="ref-Ding2020">
<p>Ding, Jiarui, Xian Adiconis, Sean K. Simmons, Monika S. Kowalczyk, Cynthia C. Hession, Nemanja D. Marjanovic, Travis K. Hughes, et al. 2020. “Systematic Comparison of Single-Cell and Single-Nucleus RNA-Sequencing Methods.” <em>Nature Biotechnology</em> 38 (6): 737–46. <a href="https://doi.org/10.1038/s41587-020-0465-8">https://doi.org/10.1038/s41587-020-0465-8</a>.</p>
</div>
<div id="ref-Freytag2018">
<p>Freytag, Saskia, Luyi Tian, Ingrid Lönnstedt, Milica Ng, and Melanie Bahlo. 2018. “Comparison of Clustering Tools in R for Medium-Sized 10x Genomics Single-Cell RNA-Sequencing Data.” <em>F1000Research</em> 7 (December): 1297. <a href="https://doi.org/10.12688/f1000research.15809.2">https://doi.org/10.12688/f1000research.15809.2</a>.</p>
</div>
<div id="ref-himes2014rna">
<p>Himes, Blanca E, Xiaofeng Jiang, Peter Wagner, Ruoxi Hu, Qiyu Wang, Barbara Klanderman, Reid M Whitaker, et al. 2014. “RNA-Seq Transcriptome Profiling Identifies Crispld2 as a Glucocorticoid Responsive Gene That Modulates Cytokine Function in Airway Smooth Muscle Cells.” <em>PloS One</em> 9 (6): e99625.</p>
</div>
<div id="ref-Karagiannis2020">
<p>Karagiannis, Tanya T., John P. Cleary, Busra Gok, Andrew J. Henderson, Nicholas G. Martin, Masanao Yajima, Elliot C. Nelson, and Christine S. Cheng. 2020. “Single Cell Transcriptomics Reveals Opioid Usage Evokes Widespread Suppression of Antiviral Gene Program.” <em>Nature Communications</em> 11 (1). <a href="https://doi.org/10.1038/s41467-020-16159-y">https://doi.org/10.1038/s41467-020-16159-y</a>.</p>
</div>
<div id="ref-law2016rna">
<p>Law, Charity W, Monther Alhamdoosh, Shian Su, Xueyi Dong, Luyi Tian, Gordon K Smyth, and Matthew E Ritchie. 2016. “RNA-Seq Analysis Is Easy as 1-2-3 with Limma, Glimma and edgeR.” <em>F1000Research</em> 5.</p>
</div>
<div id="ref-law2014voom">
<p>Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. “Voom: Precision Weights Unlock Linear Model Analysis Tools for Rna-Seq Read Counts.” <em>Genome Biology</em> 15 (2): R29.</p>
</div>
<div id="ref-liu2018integrated">
<p>Liu, Jianfang, Tara Lichtenberg, Katherine A Hoadley, Laila M Poisson, Alexander J Lazar, Andrew D Cherniack, Albert J Kovatich, et al. 2018. “An Integrated Tcga Pan-Cancer Clinical Data Resource to Drive High-Quality Survival Outcome Analytics.” <em>Cell</em> 173 (2): 400–416.</p>
</div>
<div id="ref-love2014moderated">
<p>Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for Rna-Seq Data with Deseq2.” <em>Genome Biology</em> 15 (12): 550.</p>
</div>
<div id="ref-mangiola2020tidyheatmap">
<p>Mangiola, Stefano, and Anthony T Papenfuss. 2020. “TidyHeatmap: An R Package for Modular Heatmap Production Based on Tidy Principles.” <em>Journal of Open Source Software</em> 5 (52): 2472.</p>
</div>
<div id="ref-mccarthy2012differential">
<p>McCarthy, Davis J, Yunshun Chen, and Gordon K Smyth. 2012. “Differential Expression Analysis of Multifactor Rna-Seq Experiments with Respect to Biological Variation.” <em>Nucleic Acids Research</em> 40 (10): 4288–97.</p>
</div>
<div id="ref-newman2015robust">
<p>Newman, Aaron M, Chih Long Liu, Michael R Green, Andrew J Gentles, Weiguo Feng, Yue Xu, Chuong D Hoang, Maximilian Diehn, and Ash A Alizadeh. 2015. “Robust Enumeration of Cell Subsets from Tissue Expression Profiles.” <em>Nature Methods</em> 12 (5): 453–57.</p>
</div>
<div id="ref-racle2017simultaneous">
<p>Racle, Julien, Kaat de Jonge, Petra Baumgaertner, Daniel E Speiser, and David Gfeller. 2017. “Simultaneous Enumeration of Cancer and Immune Cell Types from Bulk Tumor Gene Expression Data.” <em>Elife</em> 6: e26476.</p>
</div>
<div id="ref-robinson2010scaling">
<p>Robinson, Mark D, and Alicia Oshlack. 2010. “A Scaling Normalization Method for Differential Expression Analysis of Rna-Seq Data.” <em>Genome Biology</em> 11 (3): 1–9.</p>
</div>
<div id="ref-stuart2019comprehensive">
<p>Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck III, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. 2019. “Comprehensive Integration of Single-Cell Data.” <em>Cell</em> 177 (7): 1888–1902.</p>
</div>
<div id="ref-wickham2019welcome">
<p>Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. “Welcome to the Tidyverse.” <em>Journal of Open Source Software</em> 4 (43): 1686.</p>
</div>
<div id="ref-wickham2014tidy">
<p>Wickham, Hadley, and others. 2014. “Tidy Data.” <em>Journal of Statistical Software</em> 59 (10): 1–23.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>&lt;maria.doyle at petermac.org&gt;<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>&lt;mangiola.s at wehi.edu.au&gt;<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
